{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7a3d7a87-c836-4e8f-b855-d79b850128a5",
   "metadata": {},
   "source": [
    "A little bit of setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "868139e1-a82b-4e22-a2a2-9921132da8f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "import pandas as pd\n",
    "import json\n",
    "import shutil\n",
    "import genomitory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb3d92d6-d4a0-496d-805e-b293540e6fcf",
   "metadata": {},
   "source": [
    "## Insert SAMID and experimnet name from HITSLIMS and known sublibrary info\n",
    "- Please copy fastq path file and library list into the data folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "584b082e-f2f0-4031-9a1f-4dc2be5ad317",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAM_ID = 'SAM24449114'\n",
    "experiment_name=\"NGS5704\"\n",
    "sublibrary = \"1\" # choose among 1,2,3,4 or None, If it is not any assigned sublibraries, you have to give the cripsr index file in the data folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ebc59b1-9db8-45a1-a906-7c484e10665b",
   "metadata": {},
   "source": [
    "# section (I): Load info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99def159-ef7f-4b97-b793-7eb69345f827",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create smaple sheet from librray list comes from HITSLIMS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aca1a543-2f63-43a4-bd23-dc5c6da017bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "F = pd.read_excel(f'./data/libraryList_{experiment_name}.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9aa0ca1-e1fe-4d77-a4d4-20af65397756",
   "metadata": {},
   "outputs": [],
   "source": [
    "F = F[['LIBID','SAMID','LIBRARY_TYPE']]\n",
    "F=F.rename(columns={'LIBRARY_TYPE':'LIBRARY'})\n",
    "F.index = F.index.values\n",
    "F['index'] =F.index\n",
    "max_run=len(F)//3\n",
    "F['SAMPLE'] = F['index'].apply(lambda x:x % max_run+1) \n",
    "F = F[['LIBID', 'SAMID', 'LIBRARY', 'SAMPLE']]\n",
    "F['LIBRARY'] = F['LIBRARY'].replace({'Expression':'GEX'})\n",
    "F.to_csv(\"./data/RECRUSION_DLD1_DAY5_SAMPLE_SHEET.tsv\",sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83466c6-3622-4e7e-81cf-c87aa358aa5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this comes from hitsLims\n",
    "fastq = pd.read_excel(f'./data/{experiment_name}_ProjectFastq.xlsx').rename(columns={'SAMID':'Pool SAMID', 'Data Path':'Flowcell'})\n",
    "fastq = fastq[fastq['FASTQ QC Status'].isin(['TESTRUN', 'OK'])]\n",
    "#this comes from Russell\n",
    "exp = pd.read_csv('./data/RECRUSION_DLD1_DAY5_SAMPLE_SHEET.tsv', sep='\\t').rename(columns={'SAMID':'Pool SAMID','LIBRARY':'Library type',\n",
    "                                                                                                  'SAMPLE': 'replicate'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f86c8c3b-9cd0-4665-ad27-c77fbfbfdf2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aafc9469-9392-4fe7-bd18-47eb1c157f96",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf10501e-a4d3-4aaf-860b-bcae80423aac",
   "metadata": {},
   "source": [
    "## Calculate the size of fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9df41da1-9bc2-444e-91a2-a73f9ea0f6e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_paths = fastq[\"Flowcell\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b0ec3f-75e4-445f-9f44-8272e4945810",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_folder_size(folder_path):\n",
    "    total_size = 0\n",
    "    for dirpath, dirnames, filenames in os.walk(folder_path):\n",
    "        for filename in filenames:\n",
    "            file_path = os.path.join(dirpath, filename)\n",
    "            total_size += os.path.getsize(file_path)\n",
    "    return total_size\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ced732ae-4d18-42e9-89ee-cc43e57d97ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_size_bytes = sum(get_folder_size(fastq_path) for fastq_path in fastq_paths)\n",
    "\n",
    "# Convert bytes to GB\n",
    "combined_size_gb = combined_size_bytes / (1024 ** 3)\n",
    "\n",
    "# Print the combined size in GB\n",
    "print(f\"Combined size of folders: {combined_size_gb:.2f} GB\")\n",
    "print(f\"Number of 10x runs: {max_run}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4df64710-69f5-48e6-8ece-34fcbd9c489e",
   "metadata": {},
   "source": [
    "## Decide the number of splits based on data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45c575ee-8dab-4531-be0b-388500f64f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "N_splits = 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc184015-668e-4350-ae4a-36be8f9c1a40",
   "metadata": {},
   "source": [
    "# Section (II): Make the required directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7836cc8-185b-4a98-b26a-3406dc39240f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cur = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c9ec48cc-e8ff-4eb2-a74b-449b2f43c558",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellR, demuxH, demuxc = {},{},{}\n",
    "for i in range(N_splits):\n",
    "    folder = os.path.abspath(os.path.join(cur ,\"Split_\"+str(i+1)))\n",
    "    try:\n",
    "        os.mkdir(folder)\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "    cellR[i] = os.path.abspath(os.path.join(folder ,\"cellranger\"))\n",
    "    try:    \n",
    "        os.mkdir(cellR[i])\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "    demuxH[i] = os.path.abspath(os.path.join(folder ,\"demuxem_hashing\"))\n",
    "    try:\n",
    "        os.mkdir(demuxH[i])\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "    demuxc[i] = os.path.abspath(os.path.join(folder ,\"demuxem_crispr\"))\n",
    "    try:\n",
    "        os.mkdir(demuxc[i])\n",
    "    except FileExistsError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0a326c41-fd8f-4f39-bfea-8ca02fab973a",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = 'gs://gred-melo-carlos-lab/genomes/GRCh38_TurboGFP_3xLinker/GRCh38-2020-A_and_TurboGFP_and_3xLinker.tar.gz'\n",
    "\n",
    "library_map = {\"GEX\":'rna',\n",
    " 'HTO':'hashing',\n",
    " 'gRNA':'crispr'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "baa2a6a5-92a6-4921-b626-223b6ed8b38a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq = fastq[fastq['Pool SAMID']==SAM_ID].reset_index().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "35aa0f5e-3a1e-458b-8124-5cb0c92d8c77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>PROJECT ID</th>\n",
       "      <th>RUNID</th>\n",
       "      <th>Pool SAMID</th>\n",
       "      <th>LIBID</th>\n",
       "      <th>Flowcell</th>\n",
       "      <th>Read1</th>\n",
       "      <th>Read2</th>\n",
       "      <th>Read3</th>\n",
       "      <th>FASTQ QC Status</th>\n",
       "      <th>Library Subtype</th>\n",
       "      <th>Read Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229736</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229736_SAM24449114_S25_L005_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229736_SAM24449114_S25_L005_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>Expression</td>\n",
       "      <td>452582214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229736</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229736_SAM24449114_S25_L006_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229736_SAM24449114_S25_L006_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>Expression</td>\n",
       "      <td>456593733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229737</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229737_SAM24449114_S26_L005_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229737_SAM24449114_S26_L005_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>Expression</td>\n",
       "      <td>476188806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229737</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229737_SAM24449114_S26_L006_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229737_SAM24449114_S26_L006_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>Expression</td>\n",
       "      <td>480781247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229738</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229738_SAM24449114_S27_L005_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229738_SAM24449114_S27_L005_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>Expression</td>\n",
       "      <td>424573848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>77</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229723</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229723_SAM24449114_S32_L008_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229723_SAM24449114_S32_L008_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>gRNA</td>\n",
       "      <td>21400521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>78</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10200</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229723</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>LIB6229723_SAM24449114_S27_L004_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229723_SAM24449114_S27_L004_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>gRNA</td>\n",
       "      <td>15230018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>79</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229724</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229724_SAM24449114_S33_L007_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229724_SAM24449114_S33_L007_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>gRNA</td>\n",
       "      <td>17957975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>80</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10193</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229724</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>LIB6229724_SAM24449114_S33_L008_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229724_SAM24449114_S33_L008_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>gRNA</td>\n",
       "      <td>16520657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>81</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>R10200</td>\n",
       "      <td>SAM24449114</td>\n",
       "      <td>LIB6229724</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>LIB6229724_SAM24449114_S28_L004_R1_001.fastq.gz</td>\n",
       "      <td>LIB6229724_SAM24449114_S28_L004_R2_001.fastq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OK</td>\n",
       "      <td>gRNA</td>\n",
       "      <td>22032045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    index PROJECT ID   RUNID   Pool SAMID       LIBID  \\\n",
       "0       0    NGS5704  R10193  SAM24449114  LIB6229736   \n",
       "1       1    NGS5704  R10193  SAM24449114  LIB6229736   \n",
       "2       2    NGS5704  R10193  SAM24449114  LIB6229737   \n",
       "3       3    NGS5704  R10193  SAM24449114  LIB6229737   \n",
       "4       4    NGS5704  R10193  SAM24449114  LIB6229738   \n",
       "..    ...        ...     ...          ...         ...   \n",
       "77     77    NGS5704  R10193  SAM24449114  LIB6229723   \n",
       "78     78    NGS5704  R10200  SAM24449114  LIB6229723   \n",
       "79     79    NGS5704  R10193  SAM24449114  LIB6229724   \n",
       "80     80    NGS5704  R10193  SAM24449114  LIB6229724   \n",
       "81     81    NGS5704  R10200  SAM24449114  LIB6229724   \n",
       "\n",
       "                                          Flowcell  \\\n",
       "0   /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "1   /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "2   /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "3   /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "4   /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "..                                             ...   \n",
       "77  /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "78  /gne/data/dnaseq/processed_runs/R10200/results   \n",
       "79  /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "80  /gne/data/dnaseq/processed_runs/R10193/results   \n",
       "81  /gne/data/dnaseq/processed_runs/R10200/results   \n",
       "\n",
       "                                              Read1  \\\n",
       "0   LIB6229736_SAM24449114_S25_L005_R1_001.fastq.gz   \n",
       "1   LIB6229736_SAM24449114_S25_L006_R1_001.fastq.gz   \n",
       "2   LIB6229737_SAM24449114_S26_L005_R1_001.fastq.gz   \n",
       "3   LIB6229737_SAM24449114_S26_L006_R1_001.fastq.gz   \n",
       "4   LIB6229738_SAM24449114_S27_L005_R1_001.fastq.gz   \n",
       "..                                              ...   \n",
       "77  LIB6229723_SAM24449114_S32_L008_R1_001.fastq.gz   \n",
       "78  LIB6229723_SAM24449114_S27_L004_R1_001.fastq.gz   \n",
       "79  LIB6229724_SAM24449114_S33_L007_R1_001.fastq.gz   \n",
       "80  LIB6229724_SAM24449114_S33_L008_R1_001.fastq.gz   \n",
       "81  LIB6229724_SAM24449114_S28_L004_R1_001.fastq.gz   \n",
       "\n",
       "                                              Read2  Read3 FASTQ QC Status  \\\n",
       "0   LIB6229736_SAM24449114_S25_L005_R2_001.fastq.gz    NaN              OK   \n",
       "1   LIB6229736_SAM24449114_S25_L006_R2_001.fastq.gz    NaN              OK   \n",
       "2   LIB6229737_SAM24449114_S26_L005_R2_001.fastq.gz    NaN              OK   \n",
       "3   LIB6229737_SAM24449114_S26_L006_R2_001.fastq.gz    NaN              OK   \n",
       "4   LIB6229738_SAM24449114_S27_L005_R2_001.fastq.gz    NaN              OK   \n",
       "..                                              ...    ...             ...   \n",
       "77  LIB6229723_SAM24449114_S32_L008_R2_001.fastq.gz    NaN              OK   \n",
       "78  LIB6229723_SAM24449114_S27_L004_R2_001.fastq.gz    NaN              OK   \n",
       "79  LIB6229724_SAM24449114_S33_L007_R2_001.fastq.gz    NaN              OK   \n",
       "80  LIB6229724_SAM24449114_S33_L008_R2_001.fastq.gz    NaN              OK   \n",
       "81  LIB6229724_SAM24449114_S28_L004_R2_001.fastq.gz    NaN              OK   \n",
       "\n",
       "   Library Subtype  Read Count  \n",
       "0       Expression   452582214  \n",
       "1       Expression   456593733  \n",
       "2       Expression   476188806  \n",
       "3       Expression   480781247  \n",
       "4       Expression   424573848  \n",
       "..             ...         ...  \n",
       "77            gRNA    21400521  \n",
       "78            gRNA    15230018  \n",
       "79            gRNA    17957975  \n",
       "80            gRNA    16520657  \n",
       "81            gRNA    22032045  \n",
       "\n",
       "[82 rows x 12 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "92c84102-513c-425c-bb18-e016f307e732",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.merge(\n",
    "    fastq[['Pool SAMID', 'LIBID', 'Flowcell', 'Read1']],\n",
    "    exp[['Pool SAMID', 'LIBID', 'Library type', 'replicate']]\n",
    ")\n",
    "data['Library'] = data['Read1'].apply(lambda x:x[:22])\n",
    "data['Sample'] = data['Pool SAMID'] + '_rep' + data['replicate'].astype(str)\n",
    "data['Library type'] = data['Library type'].map(library_map)\n",
    "data = data.rename(columns={'Library type':'DataType'})\n",
    "data = data[['Library', 'Sample', 'Flowcell', 'DataType']].drop_duplicates().sort_values(by=['Sample', 'DataType'], ascending=[True, False]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "96ab7011-2067-4b39-88c4-ab9a7614e7dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Flowcell</th>\n",
       "      <th>DataType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LIB6229736_SAM24449114</td>\n",
       "      <td>SAM24449114_rep1</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LIB6229701_SAM24449114</td>\n",
       "      <td>SAM24449114_rep1</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LIB6229713_SAM24449114</td>\n",
       "      <td>SAM24449114_rep1</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LIB6229745_SAM24449114</td>\n",
       "      <td>SAM24449114_rep10</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LIB6229710_SAM24449114</td>\n",
       "      <td>SAM24449114_rep10</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LIB6229722_SAM24449114</td>\n",
       "      <td>SAM24449114_rep10</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LIB6229722_SAM24449114</td>\n",
       "      <td>SAM24449114_rep10</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LIB6229746_SAM24449114</td>\n",
       "      <td>SAM24449114_rep11</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>LIB6229711_SAM24449114</td>\n",
       "      <td>SAM24449114_rep11</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LIB6229723_SAM24449114</td>\n",
       "      <td>SAM24449114_rep11</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LIB6229723_SAM24449114</td>\n",
       "      <td>SAM24449114_rep11</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>LIB6229747_SAM24449114</td>\n",
       "      <td>SAM24449114_rep12</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>LIB6229712_SAM24449114</td>\n",
       "      <td>SAM24449114_rep12</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>LIB6229712_SAM24449114</td>\n",
       "      <td>SAM24449114_rep12</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>LIB6229724_SAM24449114</td>\n",
       "      <td>SAM24449114_rep12</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LIB6229724_SAM24449114</td>\n",
       "      <td>SAM24449114_rep12</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>LIB6229737_SAM24449114</td>\n",
       "      <td>SAM24449114_rep2</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LIB6229702_SAM24449114</td>\n",
       "      <td>SAM24449114_rep2</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>LIB6229714_SAM24449114</td>\n",
       "      <td>SAM24449114_rep2</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>LIB6229738_SAM24449114</td>\n",
       "      <td>SAM24449114_rep3</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>LIB6229703_SAM24449114</td>\n",
       "      <td>SAM24449114_rep3</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>LIB6229715_SAM24449114</td>\n",
       "      <td>SAM24449114_rep3</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>LIB6229739_SAM24449114</td>\n",
       "      <td>SAM24449114_rep4</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>LIB6229704_SAM24449114</td>\n",
       "      <td>SAM24449114_rep4</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>LIB6229716_SAM24449114</td>\n",
       "      <td>SAM24449114_rep4</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>LIB6229716_SAM24449114</td>\n",
       "      <td>SAM24449114_rep4</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>LIB6229740_SAM24449114</td>\n",
       "      <td>SAM24449114_rep5</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>LIB6229705_SAM24449114</td>\n",
       "      <td>SAM24449114_rep5</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>LIB6229717_SAM24449114</td>\n",
       "      <td>SAM24449114_rep5</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>LIB6229741_SAM24449114</td>\n",
       "      <td>SAM24449114_rep6</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>LIB6229706_SAM24449114</td>\n",
       "      <td>SAM24449114_rep6</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>LIB6229706_SAM24449114</td>\n",
       "      <td>SAM24449114_rep6</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>LIB6229718_SAM24449114</td>\n",
       "      <td>SAM24449114_rep6</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>LIB6229742_SAM24449114</td>\n",
       "      <td>SAM24449114_rep7</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>LIB6229707_SAM24449114</td>\n",
       "      <td>SAM24449114_rep7</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>LIB6229707_SAM24449114</td>\n",
       "      <td>SAM24449114_rep7</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>LIB6229719_SAM24449114</td>\n",
       "      <td>SAM24449114_rep7</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>LIB6229719_SAM24449114</td>\n",
       "      <td>SAM24449114_rep7</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>LIB6229743_SAM24449114</td>\n",
       "      <td>SAM24449114_rep8</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>LIB6229708_SAM24449114</td>\n",
       "      <td>SAM24449114_rep8</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>LIB6229708_SAM24449114</td>\n",
       "      <td>SAM24449114_rep8</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>LIB6229720_SAM24449114</td>\n",
       "      <td>SAM24449114_rep8</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>LIB6229744_SAM24449114</td>\n",
       "      <td>SAM24449114_rep9</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>LIB6229709_SAM24449114</td>\n",
       "      <td>SAM24449114_rep9</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>hashing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>LIB6229721_SAM24449114</td>\n",
       "      <td>SAM24449114_rep9</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10193/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>LIB6229721_SAM24449114</td>\n",
       "      <td>SAM24449114_rep9</td>\n",
       "      <td>/gne/data/dnaseq/processed_runs/R10200/results</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Library             Sample  \\\n",
       "0   LIB6229736_SAM24449114   SAM24449114_rep1   \n",
       "1   LIB6229701_SAM24449114   SAM24449114_rep1   \n",
       "2   LIB6229713_SAM24449114   SAM24449114_rep1   \n",
       "3   LIB6229745_SAM24449114  SAM24449114_rep10   \n",
       "4   LIB6229710_SAM24449114  SAM24449114_rep10   \n",
       "5   LIB6229722_SAM24449114  SAM24449114_rep10   \n",
       "6   LIB6229722_SAM24449114  SAM24449114_rep10   \n",
       "7   LIB6229746_SAM24449114  SAM24449114_rep11   \n",
       "8   LIB6229711_SAM24449114  SAM24449114_rep11   \n",
       "9   LIB6229723_SAM24449114  SAM24449114_rep11   \n",
       "10  LIB6229723_SAM24449114  SAM24449114_rep11   \n",
       "11  LIB6229747_SAM24449114  SAM24449114_rep12   \n",
       "12  LIB6229712_SAM24449114  SAM24449114_rep12   \n",
       "13  LIB6229712_SAM24449114  SAM24449114_rep12   \n",
       "14  LIB6229724_SAM24449114  SAM24449114_rep12   \n",
       "15  LIB6229724_SAM24449114  SAM24449114_rep12   \n",
       "16  LIB6229737_SAM24449114   SAM24449114_rep2   \n",
       "17  LIB6229702_SAM24449114   SAM24449114_rep2   \n",
       "18  LIB6229714_SAM24449114   SAM24449114_rep2   \n",
       "19  LIB6229738_SAM24449114   SAM24449114_rep3   \n",
       "20  LIB6229703_SAM24449114   SAM24449114_rep3   \n",
       "21  LIB6229715_SAM24449114   SAM24449114_rep3   \n",
       "22  LIB6229739_SAM24449114   SAM24449114_rep4   \n",
       "23  LIB6229704_SAM24449114   SAM24449114_rep4   \n",
       "24  LIB6229716_SAM24449114   SAM24449114_rep4   \n",
       "25  LIB6229716_SAM24449114   SAM24449114_rep4   \n",
       "26  LIB6229740_SAM24449114   SAM24449114_rep5   \n",
       "27  LIB6229705_SAM24449114   SAM24449114_rep5   \n",
       "28  LIB6229717_SAM24449114   SAM24449114_rep5   \n",
       "29  LIB6229741_SAM24449114   SAM24449114_rep6   \n",
       "30  LIB6229706_SAM24449114   SAM24449114_rep6   \n",
       "31  LIB6229706_SAM24449114   SAM24449114_rep6   \n",
       "32  LIB6229718_SAM24449114   SAM24449114_rep6   \n",
       "33  LIB6229742_SAM24449114   SAM24449114_rep7   \n",
       "34  LIB6229707_SAM24449114   SAM24449114_rep7   \n",
       "35  LIB6229707_SAM24449114   SAM24449114_rep7   \n",
       "36  LIB6229719_SAM24449114   SAM24449114_rep7   \n",
       "37  LIB6229719_SAM24449114   SAM24449114_rep7   \n",
       "38  LIB6229743_SAM24449114   SAM24449114_rep8   \n",
       "39  LIB6229708_SAM24449114   SAM24449114_rep8   \n",
       "40  LIB6229708_SAM24449114   SAM24449114_rep8   \n",
       "41  LIB6229720_SAM24449114   SAM24449114_rep8   \n",
       "42  LIB6229744_SAM24449114   SAM24449114_rep9   \n",
       "43  LIB6229709_SAM24449114   SAM24449114_rep9   \n",
       "44  LIB6229721_SAM24449114   SAM24449114_rep9   \n",
       "45  LIB6229721_SAM24449114   SAM24449114_rep9   \n",
       "\n",
       "                                          Flowcell DataType  \n",
       "0   /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "1   /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "2   /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "3   /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "4   /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "5   /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "6   /gne/data/dnaseq/processed_runs/R10200/results   crispr  \n",
       "7   /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "8   /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "9   /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "10  /gne/data/dnaseq/processed_runs/R10200/results   crispr  \n",
       "11  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "12  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "13  /gne/data/dnaseq/processed_runs/R10200/results  hashing  \n",
       "14  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "15  /gne/data/dnaseq/processed_runs/R10200/results   crispr  \n",
       "16  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "17  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "18  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "19  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "20  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "21  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "22  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "23  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "24  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "25  /gne/data/dnaseq/processed_runs/R10200/results   crispr  \n",
       "26  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "27  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "28  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "29  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "30  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "31  /gne/data/dnaseq/processed_runs/R10200/results  hashing  \n",
       "32  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "33  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "34  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "35  /gne/data/dnaseq/processed_runs/R10200/results  hashing  \n",
       "36  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "37  /gne/data/dnaseq/processed_runs/R10200/results   crispr  \n",
       "38  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "39  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "40  /gne/data/dnaseq/processed_runs/R10200/results  hashing  \n",
       "41  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "42  /gne/data/dnaseq/processed_runs/R10193/results      rna  \n",
       "43  /gne/data/dnaseq/processed_runs/R10193/results  hashing  \n",
       "44  /gne/data/dnaseq/processed_runs/R10193/results   crispr  \n",
       "45  /gne/data/dnaseq/processed_runs/R10200/results   crispr  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "202f2760-8dce-45f4-8ef3-069778342717",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_1/cellranger',\n",
       " 1: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_2/cellranger',\n",
       " 2: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_3/cellranger',\n",
       " 3: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_4/cellranger'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "031ac9ea-82f5-4986-a31f-4de333b85618",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_1/demuxem_hashing',\n",
       " 1: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_2/demuxem_hashing',\n",
       " 2: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_3/demuxem_hashing',\n",
       " 3: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_4/demuxem_hashing'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demuxH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f4cccae0-61f3-427d-9a59-da04dbc91ea3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_1/demuxem_crispr',\n",
       " 1: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_2/demuxem_crispr',\n",
       " 2: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_3/demuxem_crispr',\n",
       " 3: '/gstore/home/ghaffars/Cumulus/crc_ngs5704/Cumulus/Split_4/demuxem_crispr'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demuxc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "31a8c948-3950-4592-8e1d-43cd4a2d79fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SAM24449114_rep1',\n",
       " 'SAM24449114_rep10',\n",
       " 'SAM24449114_rep11',\n",
       " 'SAM24449114_rep12',\n",
       " 'SAM24449114_rep2',\n",
       " 'SAM24449114_rep3',\n",
       " 'SAM24449114_rep4',\n",
       " 'SAM24449114_rep5',\n",
       " 'SAM24449114_rep6',\n",
       " 'SAM24449114_rep7',\n",
       " 'SAM24449114_rep8',\n",
       " 'SAM24449114_rep9']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.Sample.unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e0b058-59a3-4619-b957-6f8defdeccf1",
   "metadata": {},
   "source": [
    "# Section III: Creating required data in the folders"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63d180d1-4c20-454d-a108-f01e79f3634d",
   "metadata": {},
   "source": [
    "## Split dataset into a number of splits to create some sublists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db2eb021-3211-44ca-8b6d-c9f646de2687",
   "metadata": {},
   "source": [
    "### Define a function for that reason"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6d89d83e-b987-43ca-8e5f-ff4ab410b729",
   "metadata": {},
   "outputs": [],
   "source": [
    "def divide_list (List, N_splits):\n",
    "    divison_size = len(List) // N_splits\n",
    "    remainder = len(List) % N_splits\n",
    "    \n",
    "    if remainder != 0:\n",
    "      raise Exception(\"Sorry, list should be divided to a number of splits\")\n",
    "    \n",
    "    sorted_list = sorted(List, key=lambda x:int(x[15:]))\n",
    "    final_list = [element.split('_')[1] for element in sorted_list]\n",
    "    \n",
    "    \n",
    "    divided_list=[]\n",
    "    start_idx = 0\n",
    "    for i in range(N_splits):\n",
    "        division = final_list[start_idx:start_idx+divison_size]\n",
    "        divided_list.append(division)\n",
    "        start_idx += divison_size\n",
    "        \n",
    "    return divided_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "25e47728-a8a1-4048-aee9-db1b66041be5",
   "metadata": {},
   "outputs": [],
   "source": [
    "loop = divide_list(data.Sample.unique().tolist(),N_splits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c08eec24-987d-4f1e-aae4-2e8f3fa9a1cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['rep1', 'rep2', 'rep3'],\n",
       " ['rep4', 'rep5', 'rep6'],\n",
       " ['rep7', 'rep8', 'rep9'],\n",
       " ['rep10', 'rep11', 'rep12']]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d63822a9-8147-4623-b4ea-0580c72ef1c5",
   "metadata": {},
   "source": [
    "## Write SampleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d1651e51-3e48-408c-ac69-890630788e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_sample_sheet(df, path, suffix, chem='threeprime', ref='gs://gred-melo-carlos-lab/genomes/GRCh38_TurboGFP_3xLinker/GRCh38-2020-A_and_TurboGFP_and_3xLinker.tar.gz'):\n",
    "    header = \"Library\" + \",Sample\" + \",Reference\" + \",Flowcell\" + \",Chemistry\" + \",DataType\" + \",FeatureBarcodeFile\\n\"\n",
    "    outfile = os.path.join(path, f'samplesheet_{suffix}.csv')\n",
    "    with open(outfile, 'w') as outf:\n",
    "        outf.write(f'{header}')\n",
    "        for el in range(len(df)):\n",
    "            library=df['Library'][el]\n",
    "            sample=df['Sample'][el]\n",
    "            ref=ref\n",
    "            chem=chem\n",
    "            datatype=df['DataType'][el]\n",
    "            flowcell=df['Flowcell'][el]\n",
    "            if datatype=='rna':\n",
    "                outf.write(f\"{library},{sample},{ref},{flowcell},{chem},{datatype}\\n\")\n",
    "            elif datatype=='hashing':                \n",
    "                feature='hashing_index.csv'\n",
    "                outf.write(f\"{library},{sample},{ref},{flowcell},{chem},{datatype},{feature}\\n\")\n",
    "            elif datatype=='crispr':\n",
    "                feature='crispr_index.csv'\n",
    "                outf.write(f\"{library},{sample},{ref},{flowcell},{chem},{datatype},{feature}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e8cf1f6a-9674-47e9-a76f-844d7066eca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    data_s = data[data[\"Sample\"].apply(lambda x:x.split('_')[1]).isin(loop[i])].reset_index()\n",
    "    write_sample_sheet(data_s, cellR[i], SAM_ID+\"_\"+str(i+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "447c54db-b08c-46d6-a175-47f2d69ac83a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nfor i in range(N_splits):\\n    file = os.path.join(cellR[i], f'samplesheet_SAM24432714_{i+1}.csv')\\n    os.remove(file)\\n\""
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "for i in range(N_splits):\n",
    "    file = os.path.join(cellR[i], f'samplesheet_SAM24432714_{i+1}.csv')\n",
    "    os.remove(file)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ece644c7-6577-4a2a-8e01-3eaeff88bbc5",
   "metadata": {},
   "source": [
    "## Write inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "108d594b-b91b-478d-b618-4f3c57c3d5c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    s2 = SAM_ID\n",
    "    s = SAM_ID+\"_\"+str(i+1)\n",
    "    inputs_cr = {\n",
    "    \"cellranger_workflow.input_csv_file\": f\"samplesheet_{s}.csv\",\n",
    "    \"cellranger_workflow.output_directory\": f\"gs://gred-melo-carlos-lab/{experiment_name}/{s}/\",\n",
    "    \"cellranger_workflow.run_mkfastq\": False,\n",
    "    \"cellranger_workflow.expect_cells\": 20000,\n",
    "    \"cellranger_workflow.crispr_barcode_pos\": 19,\n",
    "    \"cellranger_workflow.max_mismatch\": 0,\n",
    "    \"cellranger_workflow.write_to_dsdb\": True,\n",
    "    \"cellranger_workflow.dsdb_memory\": \"120G\",\n",
    "    \"cellranger_workflow.dsdb_disk_space\":\"5000\",\n",
    "    \"cellranger_workflow.dsdb_project_authors\": \"melocars,ghaffars,borozani\",\n",
    "    \"cellranger_workflow.dsdb_project_title\": f\"Recursion DLD1 Library4 Screen-2 (NGS5704)\",\n",
    "    \"cellranger_workflow.dsdb_project_description\": f\"Production scale screen in DLD-1 cells with 1/4 of the genome-wide CRISPR library (sub-lib-4, ~22K guides). 1 time point: Day-5. We performed 12 rxns of 10x 3' HT kit with an estimated loading of 52K cells.We performed a pilot study to test the library quality before the production-level sequencing(NGS5703). We will submit 36 libraries (12 GEX, 12 HTO and 12 sgRNA) We estimate needing - 600M per GEX, - 20M per HTO and - 50M per sgRNA libraries now that we have analyzed the QC run. This run is to add additional cells needed for analysis.\"\n",
    "        }\n",
    "    json_object = json.dumps(inputs_cr, indent=4)\n",
    "    outfile = os.path.join(cellR[i], f'inputs_{s2}.json')\n",
    "    with open(outfile, \"w\") as outfile:\n",
    "        outfile.write(json_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed3f9c87-3e16-4305-85c7-2f98461e7d24",
   "metadata": {},
   "source": [
    "## Prepare submit.sh file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b91566b7-d4aa-4b86-b846-1ecf2909fdba",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    s = SAM_ID+\"_\"+str(i+1)\n",
    "    inputs_cr = f\"\"\"\\\n",
    "#source /gstore/home/yangy197/miniconda3/bin/activate cumulus\n",
    "alto cromwell run -s cumulus.gcp.science.roche.com \\\\\n",
    "  -m cumulus/cellranger:3.0 \\\\\n",
    "  -i inputs_{SAM_ID}.json \\\\\n",
    "  -o inputs_updated.json \\\\\n",
    "  -b gs://gred-melo-carlos-lab/{experiment_name}/{s}/uploads/ \\\\\n",
    "  --no-ssl-verify \\\\\n",
    "  --no-cache \\\\\n",
    "  --profile dssc-cumulus | tee {experiment_name}.txt\n",
    "  \"\"\"\n",
    "    outfile = os.path.join(cellR[i], f'submit.sh')\n",
    "    with open(outfile, \"w\") as outfile:\n",
    "        outfile.write(inputs_cr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd5b59f7-7b74-4872-a14a-17b96a3669f2",
   "metadata": {},
   "source": [
    "## After submission, get the IDs and prepare to get metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f86651fb-2131-493f-be69-f1267b5642a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "jobids = []\n",
    "\n",
    "for i in range(N_splits):\n",
    "    outfile = os.path.join(cellR[i], f'{experiment_name}.txt')\n",
    "    with open(outfile, 'r') as file:\n",
    "            for line in file:\n",
    "                if 'Job ' in line:\n",
    "                    jobids.append(line.split(' ')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ae815a19-8bbc-4874-a901-a883ff889a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['8636e67b-9b70-474b-ad43-362cfc01117d',\n",
       " '43a564d0-d206-4245-a761-498b8dcc0c68',\n",
       " '01d7316c-2e49-4926-87fd-b05e5d669059',\n",
       " '376aa376-3ede-4641-b3f5-b52cc26c0611']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jobids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db037b2c-cd28-4117-868e-e001589485f3",
   "metadata": {},
   "source": [
    "## Write a get metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d3704a08-0141-4dd6-94a4-a9d454e991fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    outfile = os.path.join(cellR[i], f'get_metadata.sh')\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(\"#source /gstore/project/cumulus/miniconda3/bin/activate cumulus\\n\")\n",
    "        f.write(f\"alto cromwell get_metadata -s cumulus.gcp.science.roche.com --id {jobids[i]} \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d715374-5346-4ae1-a19c-08ced2e0e49a",
   "metadata": {},
   "source": [
    "## Read DSDB_ids from json metadata files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2ed55b01-c67a-4073-9763-2c68f3afa64a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dsdb = []\n",
    "\n",
    "for i in range(N_splits):\n",
    "    infile = os.path.join(cellR[i], f'{jobids[i]}.metadata.json')\n",
    "    with open(infile, 'r') as f:\n",
    "        tmp = json.load(f)\n",
    "        dsdb.append(tmp[\"outputs\"][\"cellranger_workflow.dsdb_info\"][0]['id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d8127b3b-8d9f-4837-8d00-473ea6602812",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DS000016632', 'DS000016634', 'DS000016633', 'DS000016631']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dsdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8b26d27-de18-4ce1-b454-196aeb2588a6",
   "metadata": {},
   "source": [
    "# DemuxEM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a1e2ba-37f9-429f-b5a1-3bb43b9ead90",
   "metadata": {},
   "source": [
    "## SampleSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b17a8692-b20c-473d-b2cd-43eea6e329c4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def create_demuxem_ss(data_s, barcode_datatype, path):\n",
    "    cols = ['Library', 'Sample', 'DataType']\n",
    "    df = pd.merge(data_s[data_s['DataType']=='rna'][cols], data_s[data_s['DataType']==barcode_datatype][cols], on='Sample').drop_duplicates()[\n",
    "        ['Library_x', 'Sample', 'Library_y']]\n",
    "    df.columns = ['RNA', 'OUTNAME', 'TagFile']\n",
    "    df['RNA'] = df['RNA'].apply(lambda x:path+x+'/raw_feature_bc_matrix.h5')\n",
    "    df['TagFile'] = df['TagFile'].apply(lambda x:path+x+'/'+x+f'.{barcode_datatype}.csv')\n",
    "    df['TYPE'] = 'cell-hashing'\n",
    "    return df[['OUTNAME', 'RNA', 'TagFile', 'TYPE']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "43c2f251-5f72-41c5-a0b2-a607b76eb2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    s2 = SAM_ID+\"_\"+str(i+1)\n",
    "    data_s = data[data[\"Sample\"].apply(lambda x:x.split('_')[1]).isin(loop[i])].reset_index()\n",
    "    for t in ['hashing', 'crispr']:\n",
    "            demux_s = create_demuxem_ss(data_s, t , f'gs://gred-melo-carlos-lab/{experiment_name}/{s2}/')\n",
    "            if t=='hashing':\n",
    "                outfile = os.path.join(demuxH[i], f'demux_samplesheet_{s2}.csv')\n",
    "            if t=='crispr':\n",
    "                outfile = os.path.join(demuxc[i], f'demux_samplesheet_{s2}.csv')\n",
    "            demux_s.to_csv(outfile, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3dfdfd4-ff9b-48aa-9f4b-608e87abe384",
   "metadata": {},
   "source": [
    "## jason"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "db0fa60b-8f30-4272-9957-d5e73bbc0754",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    s2 = SAM_ID\n",
    "    s = SAM_ID+\"_\"+str(i+1)\n",
    "    for t in ['hashing', 'crispr']:\n",
    "        if t == 'crispr':\n",
    "            ab_map = f\"{t}_demux_index_map.csv\"\n",
    "            outfile = os.path.join(demuxc[i], f'demux_inputs_{s2}.json')\n",
    "        if t=='hashing':\n",
    "            ab_map = f\"{t}_demux_index_map.csv\"\n",
    "            outfile = os.path.join(demuxH[i], f'demux_inputs_{s2}.json')\n",
    "        inputs_cr = {\n",
    "                \"demultiplexing.input_sample_sheet\" : f\"demux_samplesheet_{s}.csv\",\n",
    "                \"demultiplexing.output_directory\" : f\"gs://gred-melo-carlos-lab/{experiment_name}/{s}/demuxem\",\n",
    "                \"demultiplexing.genome\" : \"gs://gred-melo-carlos-lab/genomes/GRCh38_TurboGFP/GRCh38-2020-A_and_TurboGFP.tar.gz\",\n",
    "                \"demultiplexing.demuxEM_memory\":\"120G\",\n",
    "                \"demultiplexing.antibody_index_mapping_csv\" : ab_map,\n",
    "                \"demultiplexing.dsdb_id\" : dsdb[i],\n",
    "                \"demultiplexing.dsdb_project_authors\" : \"melocars,ghaffars,borozani\",\n",
    "        }\n",
    "        json_object = json.dumps(inputs_cr, indent=4)\n",
    "        with open(outfile, \"w\") as outfile:\n",
    "            outfile.write(json_object)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "751177f7-3f81-4629-b03e-09f806ab29ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nfor i in range(N_splits):\\n    s = \"SAM24432714\"\\n    file = os.path.join(demuxc[i], f\\'demux_inputs_{s}.json\\')\\n    os.remove(file)\\n'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "for i in range(N_splits):\n",
    "    s = \"SAM24432714\"\n",
    "    file = os.path.join(demuxc[i], f'demux_inputs_{s}.json')\n",
    "    os.remove(file)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743e443c-23de-420d-af9d-78b932fef705",
   "metadata": {},
   "source": [
    "## Prepare submit.sh file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a538df1-a961-40aa-ae9c-42cab928e1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    s = SAM_ID+\"_\"+str(i+1)\n",
    "    for t in ['hashing', 'crispr']:\n",
    "        inputs_cr = f\"\"\"\\\n",
    "#source /gstore/home/yangy197/miniconda3/bin/activate cumulus\n",
    "alto cromwell run -s cumulus.gcp.science.roche.com \\\\\n",
    "  -m cumulus/demultiplexing:3.0 \\\\\n",
    "  -i demux_inputs_{SAM_ID}.json \\\\\n",
    "  -o demux_inputs_updated.json \\\\\n",
    "  -b gs://gred-melo-carlos-lab/{experiment_name}/{s}/uploads/ \\\\\n",
    "  --profile dssc-cumulus \n",
    "  \"\"\"\n",
    "        if t=='hashing':\n",
    "            outfile = os.path.join(demuxH[i], f'submit.sh')\n",
    "        if t=='crispr':\n",
    "            outfile = os.path.join(demuxc[i], f'submit.sh')\n",
    "        with open(outfile, \"w\") as outfile:\n",
    "            outfile.write(inputs_cr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599aba73-8c55-4fa5-80c9-ff9033369806",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "for i in range(N_splits):\n",
    "    s2 = \"SAM24432714\"+\"_\"+str(i+1)\n",
    "    s = SAM_ID\n",
    "    for t in ['hashing', 'crispr']:\n",
    "        if t=='hashing':\n",
    "            file = os.path.join(demuxH[i], f'demux_samplesheet_{s}.csv')\n",
    "        if t=='crispr':\n",
    "            file = os.path.join(demuxc[i], f'demux_samplesheet_{s}.csv')\n",
    "        os.remove(file)\n",
    "\"\"\"       "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56395ef5-1894-4804-afe7-1dd7d152f018",
   "metadata": {},
   "source": [
    "# Section IV : Hashing and Crispr info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcba758a-d44f-4362-8c1d-c4412a55659d",
   "metadata": {},
   "source": [
    "## First revise the hashing_map file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "df8bdbe2-94ae-41ec-be1f-a191ea3c84b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map = pd.read_csv('./data/hashing.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e1dba7de-2e22-4951-b41b-9bc2e4683f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "replacement_dict = {\"SAM24434722\":\"REPLICATE_1\",\"SAM24434723\":\"REPLICATE_2\", \"SAM24434748\":\"REPLICATE_3\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "11887344-415c-43e0-8c2d-cb15196fe102",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map[\"sample\"]=hashing_map[\"sample\"].replace(replacement_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "154f4f62-97a3-46cb-9bbd-8641ac9d6bd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>HTO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTO-1</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HTO-2</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HTO-3</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HTO-4</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTO-5</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HTO-6</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HTO-7</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HTO-8</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HTO-9</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HTO-10</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HTO-11</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HTO-12</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HTO-13</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HTO-14</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HTO-15</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HTO-16</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HTO-17</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HTO-18</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HTO-19</td>\n",
       "      <td>POSITIVE_CONTROL</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index            sample cellline timepoint     HTO\n",
       "0    HTO-1       REPLICATE_1     DLD1      day5   HTO-1\n",
       "1    HTO-2       REPLICATE_1     DLD1      day5   HTO-2\n",
       "2    HTO-3       REPLICATE_1     DLD1      day5   HTO-3\n",
       "3    HTO-4       REPLICATE_2     DLD1      day5   HTO-4\n",
       "4    HTO-5       REPLICATE_2     DLD1      day5   HTO-5\n",
       "5    HTO-6       REPLICATE_2     DLD1      day5   HTO-6\n",
       "6    HTO-7       REPLICATE_3     DLD1      day5   HTO-7\n",
       "7    HTO-8       REPLICATE_3     DLD1      day5   HTO-8\n",
       "8    HTO-9       REPLICATE_3     DLD1      day5   HTO-9\n",
       "9   HTO-10       REPLICATE_1     DLD1      day5  HTO-10\n",
       "10  HTO-11       REPLICATE_1     DLD1      day5  HTO-11\n",
       "11  HTO-12       REPLICATE_1     DLD1      day5  HTO-12\n",
       "12  HTO-13       REPLICATE_2     DLD1      day5  HTO-13\n",
       "13  HTO-14       REPLICATE_2     DLD1      day5  HTO-14\n",
       "14  HTO-15       REPLICATE_2     DLD1      day5  HTO-15\n",
       "15  HTO-16       REPLICATE_3     DLD1      day5  HTO-16\n",
       "16  HTO-17       REPLICATE_3     DLD1      day5  HTO-17\n",
       "17  HTO-18       REPLICATE_3     DLD1      day5  HTO-18\n",
       "18  HTO-19  POSITIVE_CONTROL     DLD1      day5  HTO-19"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hashing_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e91cfb0d-20f5-45b3-bf16-e4928bd04426",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map[\"NGS_ID\"] = experiment_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "78e12f55-2c72-4181-93c3-c78a88605a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map[\"Biological_replicate\"]=hashing_map[\"sample\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "227ee6c1-02c8-4cd9-a23f-047e35b3387d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map[\"sublibrary\"] = f\"GMTY210:cropseq.crisprko.cas9.human.lib{sublibrary}.concerto@REVISION-3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "597cc9ce-0200-4ab8-a150-4f6c79f9b6bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map[\"gRNA_library_MOI\"] = \"0.4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "181f6933-7121-43f4-b854-78d2fd59298e",
   "metadata": {},
   "outputs": [],
   "source": [
    "replacement_dict2 = {\"REPLICATE_1\":\"SAM24439043\",\"REPLICATE_2\":\"SAM24439044\", \"REPLICATE_3\":\"SAM24439045\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9bcd2e34-8480-4bbf-8b30-836c07ff206d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map[\"sample\"]=hashing_map[\"sample\"].replace(replacement_dict2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3298b054-2d78-487b-b2ea-cb6982a57cd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sample</th>\n",
       "      <th>cellline</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>HTO</th>\n",
       "      <th>NGS_ID</th>\n",
       "      <th>Biological_replicate</th>\n",
       "      <th>sublibrary</th>\n",
       "      <th>gRNA_library_MOI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTO-1</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-1</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HTO-2</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-2</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HTO-3</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HTO-4</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-4</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTO-5</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-5</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HTO-6</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-6</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HTO-7</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-7</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HTO-8</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-8</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HTO-9</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-9</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HTO-10</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-10</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HTO-11</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-11</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HTO-12</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-12</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HTO-13</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-13</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HTO-14</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-14</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HTO-15</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-15</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HTO-16</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-16</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HTO-17</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-17</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HTO-18</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-18</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HTO-19</td>\n",
       "      <td>POSITIVE_CONTROL</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-19</td>\n",
       "      <td>NGS5704</td>\n",
       "      <td>POSITIVE_CONTROL</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib4.conce...</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index            sample cellline timepoint     HTO   NGS_ID  \\\n",
       "0    HTO-1       SAM24439043     DLD1      day5   HTO-1  NGS5704   \n",
       "1    HTO-2       SAM24439043     DLD1      day5   HTO-2  NGS5704   \n",
       "2    HTO-3       SAM24439043     DLD1      day5   HTO-3  NGS5704   \n",
       "3    HTO-4       SAM24439044     DLD1      day5   HTO-4  NGS5704   \n",
       "4    HTO-5       SAM24439044     DLD1      day5   HTO-5  NGS5704   \n",
       "5    HTO-6       SAM24439044     DLD1      day5   HTO-6  NGS5704   \n",
       "6    HTO-7       SAM24439045     DLD1      day5   HTO-7  NGS5704   \n",
       "7    HTO-8       SAM24439045     DLD1      day5   HTO-8  NGS5704   \n",
       "8    HTO-9       SAM24439045     DLD1      day5   HTO-9  NGS5704   \n",
       "9   HTO-10       SAM24439043     DLD1      day5  HTO-10  NGS5704   \n",
       "10  HTO-11       SAM24439043     DLD1      day5  HTO-11  NGS5704   \n",
       "11  HTO-12       SAM24439043     DLD1      day5  HTO-12  NGS5704   \n",
       "12  HTO-13       SAM24439044     DLD1      day5  HTO-13  NGS5704   \n",
       "13  HTO-14       SAM24439044     DLD1      day5  HTO-14  NGS5704   \n",
       "14  HTO-15       SAM24439044     DLD1      day5  HTO-15  NGS5704   \n",
       "15  HTO-16       SAM24439045     DLD1      day5  HTO-16  NGS5704   \n",
       "16  HTO-17       SAM24439045     DLD1      day5  HTO-17  NGS5704   \n",
       "17  HTO-18       SAM24439045     DLD1      day5  HTO-18  NGS5704   \n",
       "18  HTO-19  POSITIVE_CONTROL     DLD1      day5  HTO-19  NGS5704   \n",
       "\n",
       "   Biological_replicate                                         sublibrary  \\\n",
       "0           REPLICATE_1  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "1           REPLICATE_1  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "2           REPLICATE_1  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "3           REPLICATE_2  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "4           REPLICATE_2  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "5           REPLICATE_2  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "6           REPLICATE_3  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "7           REPLICATE_3  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "8           REPLICATE_3  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "9           REPLICATE_1  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "10          REPLICATE_1  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "11          REPLICATE_1  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "12          REPLICATE_2  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "13          REPLICATE_2  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "14          REPLICATE_2  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "15          REPLICATE_3  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "16          REPLICATE_3  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "17          REPLICATE_3  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "18     POSITIVE_CONTROL  GMTY210:cropseq.crisprko.cas9.human.lib4.conce...   \n",
       "\n",
       "   gRNA_library_MOI  \n",
       "0               0.4  \n",
       "1               0.4  \n",
       "2               0.4  \n",
       "3               0.4  \n",
       "4               0.4  \n",
       "5               0.4  \n",
       "6               0.4  \n",
       "7               0.4  \n",
       "8               0.4  \n",
       "9               0.4  \n",
       "10              0.4  \n",
       "11              0.4  \n",
       "12              0.4  \n",
       "13              0.4  \n",
       "14              0.4  \n",
       "15              0.4  \n",
       "16              0.4  \n",
       "17              0.4  \n",
       "18              0.4  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hashing_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "21b35665-030e-4179-ab81-af88e2b6e754",
   "metadata": {},
   "outputs": [],
   "source": [
    "hashing_map.to_csv('./data/hashing_demux_index_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b3bdff2-0775-4645-8eb4-0e550148d085",
   "metadata": {},
   "source": [
    "## Cripsr info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a963d72-d0c9-4427-86da-091d479c5890",
   "metadata": {},
   "outputs": [],
   "source": [
    "gRNA = genomitory.get_features(f\"GMTY210:cropseq.crisprko.cas9.human.lib{sublibrary}.concerto@REVISION-3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55e10804-5c24-4ad3-b812-09fe5bc83094",
   "metadata": {},
   "outputs": [],
   "source": [
    "gRNA ['index'] = gRNA ['ID']\n",
    "gRNA ['sample'] = gRNA ['ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3697b82-0abe-4ad2-816a-30356829646e",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_demux_index = gRNA[['index','sample', 'gene_symbol','class']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800fd22f-5ba3-4627-9521-242340a906fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_demux_index[crispr_demux_index.gene_symbol.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0d91d5f-e374-4c1f-839b-fa26eb974494",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_demux_index['gene_symbol'] = np.where(crispr_demux_index['gene_symbol'].isna(), crispr_demux_index['sample'].apply(lambda x:x.split('_')[0]),\n",
    "                                    crispr_demux_index['gene_symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d011b736-5940-477c-bb48-404d2a4cf608",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_demux_index[crispr_demux_index.gene_symbol.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2be105c-e257-47e8-9f42-22282d948367",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_demux_index.to_csv('./data/crispr_demux_index_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58f88690-9973-41f9-9d01-9e915572948e",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_index = gRNA[['index','spacer_19mer']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6e96bc1-d28c-435c-bb8a-d8f7f57385f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_index[\"crispr\"] = \"crispr\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2869389-a6e6-4561-b224-171cc515bed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_index=crispr_index[['spacer_19mer','index','crispr']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc3a9f44-d841-4613-a5f6-ecef75a87539",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_index.to_csv('./data/crispr_index.csv', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9691b8b4-4276-4d5f-8a0c-229ca87535c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    shutil.copy('./data/crispr_index.csv',cellR[i])\n",
    "    shutil.copy('./data/hashing_index.csv',cellR[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a3ee5b06-5972-4f7c-8b89-6129cf496965",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(N_splits):\n",
    "    shutil.copy('./data/crispr_demux_index_map.csv',demuxc[i])\n",
    "    shutil.copy('./data/hashing_demux_index_map.csv',demuxH[i])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Dev/GPuy39",
   "language": "python",
   "name": "gpuy39-dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
