{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset Crispr Hashing Harmonization: This Notebook retrieves a dataset, creates a new experiment for cripr and hashing information comnied together and then updates the dataset with new experiment of cripr and hashing harmonized "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. A little bit of set-up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Please insert your inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DSID = DatasetID\n",
    "Version = '3'   # the version of your dataset\n",
    "Version2 = '2' \n",
    "New_experiment = \"crispr_hashing_harmonized\" # A string that is a key known for the updated data saved by that that name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"..\")\n",
    "import Templates.tools.DatasetDB as DB\n",
    "import Templates.tools.scHarmony as Hproc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Instantiate a dataset object given our dataset info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "DS = DB.DATASET(DSID, DEV, title=title, description=description, name_space=name_space, organism=organism,\n",
    "                       sources=sources, tech_name=tech_name, author=author)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Access and collect Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "08:50:41 -> Collating dataset: 'DS000015686' metadata.\n",
      "08:50:41 -> Retrieving all project: 'DS000015686@3' metadata.\n",
      "\n",
      "08:50:41 -> Reading: 'DS000015686'.\n",
      "\n",
      "08:50:41 -> Reading: 'DS000015686 -> Experiment: main'.\n",
      "\n",
      "08:50:41 -> Reading: 'DS000015686 -> Experiment: main -> Alt. Exp.: crispr'.\n",
      "\n",
      "08:50:42 -> Reading: 'DS000015686 -> Experiment: main -> Alt. Exp.: hashing'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/multiassayexperiment/MultiAssayExperiment.py:127: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for group, rows in agroups:\n"
     ]
    }
   ],
   "source": [
    "adata, adatas = DS.load_dataset(Version, your_experiment, Corr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 40383 × 36603\n",
       "    obs: 'Sample', 'Barcode', 'demux_type', 'assignment', 'cellline', 'timepoint', 'HTO', 'NGS_ID', 'Biological_replicate', 'sublibrary', 'gRNA_library_MOI', 'demux_type_v2', 'assignment_v2', 'gene_symbol', 'class'\n",
       "    var: 'ID', 'Symbol', 'Type'\n",
       "    uns: 'genome', 'modality', 'uid', 'adt_summary', '.internal'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'crispr': AnnData object with n_obs × n_vars = 40383 × 21091\n",
       "     uns: '.internal'\n",
       "     layers: 'counts',\n",
       " 'hashing': AnnData object with n_obs × n_vars = 40383 × 19\n",
       "     uns: '.internal'\n",
       "     layers: 'counts'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>demux_type</th>\n",
       "      <th>assignment</th>\n",
       "      <th>cellline</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>HTO</th>\n",
       "      <th>NGS_ID</th>\n",
       "      <th>Biological_replicate</th>\n",
       "      <th>sublibrary</th>\n",
       "      <th>gRNA_library_MOI</th>\n",
       "      <th>demux_type_v2</th>\n",
       "      <th>assignment_v2</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGACCACGA</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGACCACGA</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000172538_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-7</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>FAM170B</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGAGTGGCT</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGAGTGGCT</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000213337_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-1</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>ANKRD39</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGAGTGTTA</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGAGTGTTA</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000128309_4</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-8</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>MPST</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGCACTCAT</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGCACTCAT</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000184445_4</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-13</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>KNTC1</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGCATTGTC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGCATTGTC</td>\n",
       "      <td>doublet</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-6</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCCATTTAC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCCATTTAC</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCCCATTCG</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCCCATTCG</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000165495_3</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>doublet</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>PKNOX2</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTGACGCG</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCTGACGCG</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000164989_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>CCDC171</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTTAATCC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCTTAATCC</td>\n",
       "      <td>doublet</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>doublet</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTTCTAAC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCTTCTAAC</td>\n",
       "      <td>singlet</td>\n",
       "      <td>NTC_671</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-12</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>NTC</td>\n",
       "      <td>ntc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40383 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sample           Barcode  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA  SAM24439046_rep1  AAACCCAAGACCACGA   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT  SAM24439046_rep1  AAACCCAAGAGTGGCT   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA  SAM24439046_rep1  AAACCCAAGAGTGTTA   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT  SAM24439046_rep1  AAACCCAAGCACTCAT   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  SAM24439046_rep1  AAACCCAAGCATTGTC   \n",
       "...                                             ...               ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC  SAM24439046_rep1  TTTGTTGTCCATTTAC   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG  SAM24439046_rep1  TTTGTTGTCCCATTCG   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG  SAM24439046_rep1  TTTGTTGTCTGACGCG   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  SAM24439046_rep1  TTTGTTGTCTTAATCC   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC  SAM24439046_rep1  TTTGTTGTCTTCTAAC   \n",
       "\n",
       "                                  demux_type  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC    doublet   \n",
       "...                                      ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC    unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG    singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG    singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC    doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC    singlet   \n",
       "\n",
       "                                                            assignment  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                    ENSG00000172538_2   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                    ENSG00000213337_2   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                    ENSG00000128309_4   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                    ENSG00000184445_4   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3   \n",
       "...                                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                    ENSG00000165495_3   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                    ENSG00000164989_2   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                              NTC_671   \n",
       "\n",
       "                                        cellline      timepoint  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC           DLD1           day5   \n",
       "...                                          ...            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC           DLD1           day5   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG           DLD1           day5   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC           DLD1           day5   \n",
       "\n",
       "                                             HTO         NGS_ID  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          HTO-7        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          HTO-1        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          HTO-8        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT         HTO-13        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          HTO-6        NGS5445   \n",
       "...                                          ...            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          HTO-3        NGS5445   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          HTO-3        NGS5445   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC         HTO-12        NGS5445   \n",
       "\n",
       "                                  Biological_replicate  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          REPLICATE_3   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          REPLICATE_1   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          REPLICATE_3   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT          REPLICATE_2   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          REPLICATE_2   \n",
       "...                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          REPLICATE_1   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG         HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          REPLICATE_1   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC        HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC          REPLICATE_1   \n",
       "\n",
       "                                                                          sublibrary  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "...                                                                              ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                                       HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC                                      HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "\n",
       "                                  gRNA_library_MOI demux_type_v2  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC              0.4       singlet   \n",
       "...                                            ...           ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC              0.4       singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG     HTO-1,HTO-18       doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG              0.4       singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC    HTO-14,HTO-16       doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC              0.4       singlet   \n",
       "\n",
       "                                   assignment_v2  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA    SAM24439045   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT    SAM24439043   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA    SAM24439045   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT    SAM24439044   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC    SAM24439044   \n",
       "...                                          ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC    SAM24439043   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG    SAM24439043   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC    SAM24439043   \n",
       "\n",
       "                                                           gene_symbol  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                              FAM170B   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                              ANKRD39   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                                 MPST   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                                KNTC1   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3   \n",
       "...                                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                               PKNOX2   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                              CCDC171   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                                  NTC   \n",
       "\n",
       "                                                                 class  \n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                               target  \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                               target  \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                               target  \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                               target  \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3  \n",
       "...                                                                ...  \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown  \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                               target  \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                               target  \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4  \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                                  ntc  \n",
       "\n",
       "[40383 rows x 15 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['3xLinker_3xLinker', 'GRCh38-2020-A_ENSG00000243485',\n",
       "       'GRCh38-2020-A_ENSG00000237613', 'GRCh38-2020-A_ENSG00000186092',\n",
       "       'GRCh38-2020-A_ENSG00000238009', 'GRCh38-2020-A_ENSG00000239945',\n",
       "       'GRCh38-2020-A_ENSG00000239906', 'GRCh38-2020-A_ENSG00000241860',\n",
       "       'GRCh38-2020-A_ENSG00000241599', 'GRCh38-2020-A_ENSG00000286448',\n",
       "       ...\n",
       "       'GRCh38-2020-A_ENSG00000275869', 'GRCh38-2020-A_ENSG00000273554',\n",
       "       'GRCh38-2020-A_ENSG00000278782', 'GRCh38-2020-A_ENSG00000277761',\n",
       "       'GRCh38-2020-A_ENSG00000277836', 'GRCh38-2020-A_ENSG00000278633',\n",
       "       'GRCh38-2020-A_ENSG00000276017', 'GRCh38-2020-A_ENSG00000278817',\n",
       "       'GRCh38-2020-A_ENSG00000277196', 'TurboGFP_TurboGFP'],\n",
       "      dtype='object', length=36603)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3xLinker_3xLinker</th>\n",
       "      <td>3xLinker_3xLinker</td>\n",
       "      <td>3xLinker_3xLinker</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38-2020-A_ENSG00000243485</th>\n",
       "      <td>GRCh38-2020-A_ENSG00000243485</td>\n",
       "      <td>GRCh38-2020-A_MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38-2020-A_ENSG00000237613</th>\n",
       "      <td>GRCh38-2020-A_ENSG00000237613</td>\n",
       "      <td>GRCh38-2020-A_FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38-2020-A_ENSG00000186092</th>\n",
       "      <td>GRCh38-2020-A_ENSG00000186092</td>\n",
       "      <td>GRCh38-2020-A_OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRCh38-2020-A_ENSG00000238009</th>\n",
       "      <td>GRCh38-2020-A_ENSG00000238009</td>\n",
       "      <td>GRCh38-2020-A_AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          ID  \\\n",
       "3xLinker_3xLinker                          3xLinker_3xLinker   \n",
       "GRCh38-2020-A_ENSG00000243485  GRCh38-2020-A_ENSG00000243485   \n",
       "GRCh38-2020-A_ENSG00000237613  GRCh38-2020-A_ENSG00000237613   \n",
       "GRCh38-2020-A_ENSG00000186092  GRCh38-2020-A_ENSG00000186092   \n",
       "GRCh38-2020-A_ENSG00000238009  GRCh38-2020-A_ENSG00000238009   \n",
       "\n",
       "                                                  Symbol             Type  \n",
       "3xLinker_3xLinker                      3xLinker_3xLinker  Gene Expression  \n",
       "GRCh38-2020-A_ENSG00000243485  GRCh38-2020-A_MIR1302-2HG  Gene Expression  \n",
       "GRCh38-2020-A_ENSG00000237613      GRCh38-2020-A_FAM138A  Gene Expression  \n",
       "GRCh38-2020-A_ENSG00000186092        GRCh38-2020-A_OR4F5  Gene Expression  \n",
       "GRCh38-2020-A_ENSG00000238009   GRCh38-2020-A_AL627309.1  Gene Expression  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<40383x36603 sparse matrix of type '<class 'numpy.int32'>'\n",
       "\twith 40520263 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Modify the anndata using a function (any operations you want)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = adata.obs.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>demux_type</th>\n",
       "      <th>assignment</th>\n",
       "      <th>cellline</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>HTO</th>\n",
       "      <th>NGS_ID</th>\n",
       "      <th>Biological_replicate</th>\n",
       "      <th>sublibrary</th>\n",
       "      <th>gRNA_library_MOI</th>\n",
       "      <th>demux_type_v2</th>\n",
       "      <th>assignment_v2</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGACCACGA</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGACCACGA</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000172538_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-7</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>FAM170B</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGAGTGGCT</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGAGTGGCT</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000213337_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-1</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>ANKRD39</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGAGTGTTA</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGAGTGTTA</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000128309_4</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-8</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>MPST</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGCACTCAT</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGCACTCAT</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000184445_4</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-13</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>KNTC1</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGCATTGTC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>AAACCCAAGCATTGTC</td>\n",
       "      <td>doublet</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-6</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCCATTTAC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCCATTTAC</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCCCATTCG</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCCCATTCG</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000165495_3</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>doublet</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>PKNOX2</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTGACGCG</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCTGACGCG</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000164989_2</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>CCDC171</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTTAATCC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCTTAATCC</td>\n",
       "      <td>doublet</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>doublet</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTTCTAAC</th>\n",
       "      <td>SAM24439046_rep1</td>\n",
       "      <td>TTTGTTGTCTTCTAAC</td>\n",
       "      <td>singlet</td>\n",
       "      <td>NTC_671</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-12</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>NTC</td>\n",
       "      <td>ntc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40383 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sample           Barcode  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA  SAM24439046_rep1  AAACCCAAGACCACGA   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT  SAM24439046_rep1  AAACCCAAGAGTGGCT   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA  SAM24439046_rep1  AAACCCAAGAGTGTTA   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT  SAM24439046_rep1  AAACCCAAGCACTCAT   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  SAM24439046_rep1  AAACCCAAGCATTGTC   \n",
       "...                                             ...               ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC  SAM24439046_rep1  TTTGTTGTCCATTTAC   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG  SAM24439046_rep1  TTTGTTGTCCCATTCG   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG  SAM24439046_rep1  TTTGTTGTCTGACGCG   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  SAM24439046_rep1  TTTGTTGTCTTAATCC   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC  SAM24439046_rep1  TTTGTTGTCTTCTAAC   \n",
       "\n",
       "                                  demux_type  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT    singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC    doublet   \n",
       "...                                      ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC    unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG    singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG    singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC    doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC    singlet   \n",
       "\n",
       "                                                            assignment  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                    ENSG00000172538_2   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                    ENSG00000213337_2   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                    ENSG00000128309_4   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                    ENSG00000184445_4   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3   \n",
       "...                                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                    ENSG00000165495_3   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                    ENSG00000164989_2   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                              NTC_671   \n",
       "\n",
       "                                        cellline      timepoint  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC           DLD1           day5   \n",
       "...                                          ...            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC           DLD1           day5   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG           DLD1           day5   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC           DLD1           day5   \n",
       "\n",
       "                                             HTO         NGS_ID  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          HTO-7        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          HTO-1        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          HTO-8        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT         HTO-13        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          HTO-6        NGS5445   \n",
       "...                                          ...            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          HTO-3        NGS5445   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          HTO-3        NGS5445   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC         HTO-12        NGS5445   \n",
       "\n",
       "                                  Biological_replicate  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          REPLICATE_3   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          REPLICATE_1   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          REPLICATE_3   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT          REPLICATE_2   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          REPLICATE_2   \n",
       "...                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          REPLICATE_1   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG         HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          REPLICATE_1   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC        HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC          REPLICATE_1   \n",
       "\n",
       "                                                                          sublibrary  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "...                                                                              ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                                       HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC                                      HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "\n",
       "                                  gRNA_library_MOI demux_type_v2  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT              0.4       singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC              0.4       singlet   \n",
       "...                                            ...           ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC              0.4       singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG     HTO-1,HTO-18       doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG              0.4       singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC    HTO-14,HTO-16       doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC              0.4       singlet   \n",
       "\n",
       "                                   assignment_v2  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA    SAM24439045   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT    SAM24439043   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA    SAM24439045   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT    SAM24439044   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC    SAM24439044   \n",
       "...                                          ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC    SAM24439043   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG    SAM24439043   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC    SAM24439043   \n",
       "\n",
       "                                                           gene_symbol  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                              FAM170B   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                              ANKRD39   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                                 MPST   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                                KNTC1   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3   \n",
       "...                                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                               PKNOX2   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                              CCDC171   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                                  NTC   \n",
       "\n",
       "                                                                 class  \n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                               target  \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                               target  \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                               target  \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                               target  \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3  \n",
       "...                                                                ...  \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown  \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                               target  \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                               target  \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4  \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                                  ntc  \n",
       "\n",
       "[40383 rows x 15 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = tmp.rename(columns={\"demux_type\":\"DemuxType_crispr\",\"assignment\":\"DemuxAssignment_crispr\",\n",
    "                          \"demux_type_v2\":\"DemuxType_hashing\",\"assignment_v2\":\"DemuxAssignment_hashing\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp[\"10Xrun\"] = tmp [\"Sample\"].str.split('_').str[1]\n",
    "tmp[\"10Xrun\"] = tmp [\"10Xrun\"].str.extract('(\\d+)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp[\"Sample\"] = tmp [\"Sample\"].str.split('_').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = tmp[['Sample', 'Barcode', 'DemuxType_crispr', 'DemuxAssignment_crispr',\n",
    "       'DemuxType_hashing', 'DemuxAssignment_hashing', 'cellline', 'timepoint','HTO', 'NGS_ID',\n",
    "       'Biological_replicate', '10Xrun', 'sublibrary', 'gRNA_library_MOI',\n",
    "       'gene_symbol', 'class']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>DemuxType_crispr</th>\n",
       "      <th>DemuxAssignment_crispr</th>\n",
       "      <th>DemuxType_hashing</th>\n",
       "      <th>DemuxAssignment_hashing</th>\n",
       "      <th>cellline</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>HTO</th>\n",
       "      <th>NGS_ID</th>\n",
       "      <th>Biological_replicate</th>\n",
       "      <th>10Xrun</th>\n",
       "      <th>sublibrary</th>\n",
       "      <th>gRNA_library_MOI</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGACCACGA</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>AAACCCAAGACCACGA</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000172538_2</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-7</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>FAM170B</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGAGTGGCT</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>AAACCCAAGAGTGGCT</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000213337_2</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-1</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>ANKRD39</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGAGTGTTA</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>AAACCCAAGAGTGTTA</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000128309_4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439045</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-8</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_3</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>MPST</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGCACTCAT</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>AAACCCAAGCACTCAT</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000184445_4</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-13</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>KNTC1</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-AAACCCAAGCATTGTC</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>AAACCCAAGCATTGTC</td>\n",
       "      <td>doublet</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439044</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-6</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_2</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "      <td>ENSG00000197747_3,ENSG00000022355_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCCATTTAC</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>TTTGTTGTCCATTTAC</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCCCATTCG</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>TTTGTTGTCCCATTCG</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000165495_3</td>\n",
       "      <td>doublet</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>1</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>HTO-1,HTO-18</td>\n",
       "      <td>PKNOX2</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTGACGCG</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>TTTGTTGTCTGACGCG</td>\n",
       "      <td>singlet</td>\n",
       "      <td>ENSG00000164989_2</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-3</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>CCDC171</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTTAATCC</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>TTTGTTGTCTTAATCC</td>\n",
       "      <td>doublet</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "      <td>doublet</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>1</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>HTO-14,HTO-16</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "      <td>ENSG00000100302_3,ENSG00000158623_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAM24439046_rep1-TTTGTTGTCTTCTAAC</th>\n",
       "      <td>SAM24439046</td>\n",
       "      <td>TTTGTTGTCTTCTAAC</td>\n",
       "      <td>singlet</td>\n",
       "      <td>NTC_671</td>\n",
       "      <td>singlet</td>\n",
       "      <td>SAM24439043</td>\n",
       "      <td>DLD1</td>\n",
       "      <td>day5</td>\n",
       "      <td>HTO-12</td>\n",
       "      <td>NGS5445</td>\n",
       "      <td>REPLICATE_1</td>\n",
       "      <td>1</td>\n",
       "      <td>GMTY210:cropseq.crisprko.cas9.human.lib2.conce...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>NTC</td>\n",
       "      <td>ntc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40383 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Sample           Barcode  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA  SAM24439046  AAACCCAAGACCACGA   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT  SAM24439046  AAACCCAAGAGTGGCT   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA  SAM24439046  AAACCCAAGAGTGTTA   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT  SAM24439046  AAACCCAAGCACTCAT   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  SAM24439046  AAACCCAAGCATTGTC   \n",
       "...                                        ...               ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC  SAM24439046  TTTGTTGTCCATTTAC   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG  SAM24439046  TTTGTTGTCCCATTCG   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG  SAM24439046  TTTGTTGTCTGACGCG   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  SAM24439046  TTTGTTGTCTTAATCC   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC  SAM24439046  TTTGTTGTCTTCTAAC   \n",
       "\n",
       "                                  DemuxType_crispr  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          singlet   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT          singlet   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          doublet   \n",
       "...                                            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG          singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          singlet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC          doublet   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC          singlet   \n",
       "\n",
       "                                                DemuxAssignment_crispr  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                    ENSG00000172538_2   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                    ENSG00000213337_2   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                    ENSG00000128309_4   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                    ENSG00000184445_4   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3   \n",
       "...                                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                    ENSG00000165495_3   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                    ENSG00000164989_2   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                              NTC_671   \n",
       "\n",
       "                                  DemuxType_hashing DemuxAssignment_hashing  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA           singlet             SAM24439045   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT           singlet             SAM24439043   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA           singlet             SAM24439045   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT           singlet             SAM24439044   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC           singlet             SAM24439044   \n",
       "...                                             ...                     ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC           singlet             SAM24439043   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG           doublet            HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG           singlet             SAM24439043   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC           doublet           HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC           singlet             SAM24439043   \n",
       "\n",
       "                                        cellline      timepoint  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT           DLD1           day5   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC           DLD1           day5   \n",
       "...                                          ...            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC           DLD1           day5   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG           DLD1           day5   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC           DLD1           day5   \n",
       "\n",
       "                                             HTO         NGS_ID  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          HTO-7        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          HTO-1        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          HTO-8        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT         HTO-13        NGS5445   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          HTO-6        NGS5445   \n",
       "...                                          ...            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          HTO-3        NGS5445   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG   HTO-1,HTO-18   HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          HTO-3        NGS5445   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  HTO-14,HTO-16  HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC         HTO-12        NGS5445   \n",
       "\n",
       "                                  Biological_replicate 10Xrun  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA          REPLICATE_3      1   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT          REPLICATE_1      1   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA          REPLICATE_3      1   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT          REPLICATE_2      1   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC          REPLICATE_2      1   \n",
       "...                                                ...    ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC          REPLICATE_1      1   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG         HTO-1,HTO-18      1   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG          REPLICATE_1      1   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC        HTO-14,HTO-16      1   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC          REPLICATE_1      1   \n",
       "\n",
       "                                                                          sublibrary  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "...                                                                              ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                                       HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC                                      HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC  GMTY210:cropseq.crisprko.cas9.human.lib2.conce...   \n",
       "\n",
       "                                  gRNA_library_MOI  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA              0.4   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT              0.4   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA              0.4   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT              0.4   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC              0.4   \n",
       "...                                            ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC              0.4   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG     HTO-1,HTO-18   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG              0.4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC    HTO-14,HTO-16   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC              0.4   \n",
       "\n",
       "                                                           gene_symbol  \\\n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                              FAM170B   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                              ANKRD39   \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                                 MPST   \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                                KNTC1   \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3   \n",
       "...                                                                ...   \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown   \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                               PKNOX2   \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                              CCDC171   \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4   \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                                  NTC   \n",
       "\n",
       "                                                                 class  \n",
       "SAM24439046_rep1-AAACCCAAGACCACGA                               target  \n",
       "SAM24439046_rep1-AAACCCAAGAGTGGCT                               target  \n",
       "SAM24439046_rep1-AAACCCAAGAGTGTTA                               target  \n",
       "SAM24439046_rep1-AAACCCAAGCACTCAT                               target  \n",
       "SAM24439046_rep1-AAACCCAAGCATTGTC  ENSG00000197747_3,ENSG00000022355_3  \n",
       "...                                                                ...  \n",
       "SAM24439046_rep1-TTTGTTGTCCATTTAC                              unknown  \n",
       "SAM24439046_rep1-TTTGTTGTCCCATTCG                               target  \n",
       "SAM24439046_rep1-TTTGTTGTCTGACGCG                               target  \n",
       "SAM24439046_rep1-TTTGTTGTCTTAATCC  ENSG00000100302_3,ENSG00000158623_4  \n",
       "SAM24439046_rep1-TTTGTTGTCTTCTAAC                                  ntc  \n",
       "\n",
       "[40383 rows x 16 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_updated = adata.copy()\n",
    "adatas_updated = adatas.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 40383 × 36603\n",
       "    obs: 'Sample', 'Barcode', 'demux_type', 'assignment', 'cellline', 'timepoint', 'HTO', 'NGS_ID', 'Biological_replicate', 'sublibrary', 'gRNA_library_MOI', 'demux_type_v2', 'assignment_v2', 'gene_symbol', 'class'\n",
       "    var: 'ID', 'Symbol', 'Type'\n",
       "    uns: 'genome', 'modality', 'uid', 'adt_summary', '.internal'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_updated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'crispr': AnnData object with n_obs × n_vars = 40383 × 21091\n",
       "     uns: '.internal'\n",
       "     layers: 'counts',\n",
       " 'hashing': AnnData object with n_obs × n_vars = 40383 × 19\n",
       "     uns: '.internal'\n",
       "     layers: 'counts'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas_updated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_updated.obs = tmp.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 40383 × 36603\n",
       "    obs: 'Sample', 'Barcode', 'DemuxType_crispr', 'DemuxAssignment_crispr', 'DemuxType_hashing', 'DemuxAssignment_hashing', 'cellline', 'timepoint', 'HTO', 'NGS_ID', 'Biological_replicate', '10Xrun', 'sublibrary', 'gRNA_library_MOI', 'gene_symbol', 'class'\n",
       "    var: 'ID', 'Symbol', 'Type'\n",
       "    uns: 'genome', 'modality', 'uid', 'adt_summary', '.internal'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_updated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'crispr': AnnData object with n_obs × n_vars = 40383 × 21091\n",
       "     uns: '.internal'\n",
       "     layers: 'counts',\n",
       " 'hashing': AnnData object with n_obs × n_vars = 40383 × 19\n",
       "     uns: '.internal'\n",
       "     layers: 'counts'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas_updated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Upload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "09:00:02 -> Collating dataset: 'DS000015686' metadata.\n",
      "09:00:02 -> Retrieving all project: 'DS000015686@1' metadata.\n",
      "\n",
      "09:00:03 -> Reading: 'DS000015686'.\n",
      "\n",
      "09:00:03 -> Reading: 'DS000015686 -> Experiment: main'.\n",
      "\n",
      "09:00:03 -> Reading: 'DS000015686 -> Experiment: main -> Alt. Exp.: crispr'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DS000015686 -> Experiment: main -> Alt. Exp.: crispr -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 551kB/s]\n",
      "DS000015686 -> Experiment: main -> Alt. Exp.: crispr -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 2.16M/2.16M [00:06<00:00, 338kB/s] \n",
      "DS000015686 -> Experiment: main -> Alt. Exp.: crispr -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 168k/168k [00:00<00:00, 55.8MB/s]\n",
      "DS000015686 -> Experiment: main -> Alt. Exp.: crispr -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 51.8k/51.8k [00:00<00:00, 29.0MB/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "09:00:15 -> Reading: 'DS000015686 -> Experiment: main -> Alt. Exp.: hashing'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DS000015686 -> Experiment: main -> Alt. Exp.: hashing -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 532kB/s]\n",
      "DS000015686 -> Experiment: main -> Alt. Exp.: hashing -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 2.46M/2.46M [00:01<00:00, 1.49MB/s]\n",
      "DS000015686 -> Experiment: main -> Alt. Exp.: hashing -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 168k/168k [00:00<00:00, 2.20MB/s]\n",
      "DS000015686 -> Experiment: main -> Alt. Exp.: hashing -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 90.0/90.0 [00:00<00:00, 65.8kB/s]\n",
      "DS000015686 -> Experiment: main -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 610kB/s]\n",
      "DS000015686 -> Experiment: main -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 243M/243M [00:14<00:00, 16.3MB/s] \n",
      "DS000015686 -> Experiment: main -> Column Data -> Data Frame -> Column: 'Sample' -> Data Frame: 100%|████████████████████████████████████████| 48.0/48.0 [00:00<00:00, 35.2kB/s]\n",
      "DS000015686 -> Experiment: main -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 266k/266k [00:00<00:00, 77.6MB/s]\n",
      "DS000015686 -> Experiment: main -> Other Data -> Basic List: 100%|████████████████████████████████████████| 474/474 [00:00<00:00, 334kB/s]\n",
      "DS000015686 -> Experiment: main -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 408k/408k [00:00<00:00, 16.4MB/s]\n",
      "DS000015686 -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 78.0/78.0 [00:00<00:00, 56.2kB/s]\n",
      "DS000015686 -> Sample Mapping: 100%|████████████████████████████████████████| 279k/279k [00:00<00:00, 7.86MB/s]\n",
      "/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/multiassayexperiment/MultiAssayExperiment.py:127: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for group, rows in agroups:\n",
      "\n",
      "PYDSDB:INFO: Experiment already contained metadata, overwriting.\n",
      "\n",
      "PYDSDB:INFO: Experiment already contained metadata, overwriting.\n",
      "\n",
      "PYDSDB:INFO: Experiment already contained metadata, overwriting.\n",
      "/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/multiassayexperiment/MultiAssayExperiment.py:127: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for group, rows in agroups:\n",
      "/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/multiassayexperiment/MultiAssayExperiment.py:127: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for group, rows in agroups:\n"
     ]
    },
    {
     "ename": "PYDSDBError",
     "evalue": "For column: 'DemuxType_hashing' with row 'object' dtype row values have more than one type (excluding 'None'): '{<class 'float'>, <class 'str'>}', this is not supported.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPYDSDBError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[30], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m DS_test\u001b[38;5;241m==\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m:\n\u001b[0;32m----> 2\u001b[0m     dsid, version \u001b[38;5;241m=\u001b[39m \u001b[43mDS\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupdate_dataset\u001b[49m\u001b[43m(\u001b[49m\u001b[43mNew_experiment\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madata_updated\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madatas_updated\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtest\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:    \n\u001b[1;32m      4\u001b[0m     dsid, version \u001b[38;5;241m=\u001b[39m DS\u001b[38;5;241m.\u001b[39mupdate_another_dataset(test_DatasetID, \u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;241m1\u001b[39m, New_experiment, adata_updated, adatas_updated, test\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/Cumulus/crc_ngs5445/crop-seq-pipeline/src/Templates/../Templates/tools/DatasetDB.py:224\u001b[0m, in \u001b[0;36mDATASET.update_dataset\u001b[0;34m(self, new_experiment, adata_updated, adatas_updated, test, Corr)\u001b[0m\n\u001b[1;32m    217\u001b[0m     new_mae \u001b[38;5;241m=\u001b[39m mae\u001b[38;5;241m.\u001b[39mMultiAssayExperiment(\n\u001b[1;32m    218\u001b[0m     {\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mexperiments,new_experiment: tse},\n\u001b[1;32m    219\u001b[0m     column_data,\n\u001b[1;32m    220\u001b[0m     sample_map,\n\u001b[1;32m    221\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmetadata\n\u001b[1;32m    222\u001b[0m     )\n\u001b[1;32m    223\u001b[0m \u001b[38;5;66;03m# Initialize the upload\u001b[39;00m\n\u001b[0;32m--> 224\u001b[0m upload \u001b[38;5;241m=\u001b[39m \u001b[43mpydsdb\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mUpload\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnew_mae\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    225\u001b[0m \u001b[38;5;66;03m# Set Permissions\u001b[39;00m\n\u001b[1;32m    226\u001b[0m permissions \u001b[38;5;241m=\u001b[39m pydsdb\u001b[38;5;241m.\u001b[39mcreate_permissions_info()\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/upload.py:125\u001b[0m, in \u001b[0;36mUpload.__init__\u001b[0;34m(self, dataset, staging_dir, descriptor, file_size_limit, overwrite)\u001b[0m\n\u001b[1;32m    123\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_staging_dir \u001b[38;5;241m=\u001b[39m staging_dir\n\u001b[1;32m    124\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 125\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_staging_dir \u001b[38;5;241m=\u001b[39m \u001b[43mDatasetStager\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    126\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdataset\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdescriptor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfile_size_limit\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moverwrite\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstaging_dir\u001b[49m\n\u001b[1;32m    127\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstage\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_permissions_info: Optional[PermissionsInfo] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    131\u001b[0m     dataset \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    132\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m dataset\u001b[38;5;241m.\u001b[39mmetadata \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m    133\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m INTERNAL \u001b[38;5;129;01min\u001b[39;00m dataset\u001b[38;5;241m.\u001b[39mmetadata  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m    134\u001b[0m ):\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/dataset/generic.py:164\u001b[0m, in \u001b[0;36mDatasetStager.stage\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    156\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m update_info\u001b[38;5;241m.\u001b[39mupdate_ready \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m:\n\u001b[1;32m    157\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PYDSDBError(\n\u001b[1;32m    158\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot stage this dataset as it is missing data due to \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    159\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpartially supported resource types\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    160\u001b[0m     )\n\u001b[1;32m    162\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_write_metadata(\n\u001b[1;32m    163\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_paths(\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdataset\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m2\u001b[39m],\n\u001b[0;32m--> 164\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[43m)\u001b[49m,\n\u001b[1;32m    165\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mloads(update_info\u001b[38;5;241m.\u001b[39mmetadata)[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m$schema\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    166\u001b[0m )\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_staging_dir\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/dataset/stagers/dataset.py:199\u001b[0m, in \u001b[0;36mDatasetStager._make_metadata\u001b[0;34m(self, update_info)\u001b[0m\n\u001b[1;32m    193\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PYDSDBError(\n\u001b[1;32m    194\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDataset: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_descriptor\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m metadata could not be parsed!\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    195\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mexpt\u001b[39;00m\n\u001b[1;32m    197\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_staging_dir\u001b[38;5;241m.\u001b[39mjoinpath(OLD_PATHS)\u001b[38;5;241m.\u001b[39mwrite_text(dumps(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mold_paths))\n\u001b[0;32m--> 199\u001b[0m metadata\u001b[38;5;241m.\u001b[39mdataset \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_stage_dataset_components\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    200\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdataset\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    201\u001b[0m \u001b[43m    \u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumn_data\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    202\u001b[0m \u001b[43m    \u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msample_mapping\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    203\u001b[0m \u001b[43m    \u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexp_nums\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    204\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    206\u001b[0m metadata\u001b[38;5;241m.\u001b[39mis_child \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_child\n\u001b[1;32m    208\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m metadata\u001b[38;5;241m.\u001b[39mdict()\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/dataset/stagers/dataset.py:127\u001b[0m, in \u001b[0;36mDatasetStager._stage_dataset_components\u001b[0;34m(self, metadata, orig_column_data, orig_sample_mapping, exp_nums)\u001b[0m\n\u001b[1;32m    122\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    123\u001b[0m         exp_stager \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_child_experiment_stager(\n\u001b[1;32m    124\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_se_stager, exp, \u001b[38;5;241m*\u001b[39mexp_args\n\u001b[1;32m    125\u001b[0m         )\n\u001b[0;32m--> 127\u001b[0m     exps\u001b[38;5;241m.\u001b[39mappend(\u001b[43mexp_stager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstage\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    128\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mold_paths\u001b[38;5;241m.\u001b[39mextend(exp_stager\u001b[38;5;241m.\u001b[39mold_paths)\n\u001b[1;32m    130\u001b[0m metadata\u001b[38;5;241m.\u001b[39mexperiments \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_combine_named_resources(\n\u001b[1;32m    131\u001b[0m     metadata\u001b[38;5;241m.\u001b[39mexperiments, exps\n\u001b[1;32m    132\u001b[0m )\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/experiments/generic.py:211\u001b[0m, in \u001b[0;36mExperimentStager.stage\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    205\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m update_info\u001b[38;5;241m.\u001b[39mupdate_ready \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m:\n\u001b[1;32m    206\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PYDSDBError(\n\u001b[1;32m    207\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot stage this experiment as it is missing data due to \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    208\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpartially supported resource types\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    209\u001b[0m     )\n\u001b[0;32m--> 211\u001b[0m metadata \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m metadata_path, metadata[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpath\u001b[39m\u001b[38;5;124m\"\u001b[39m], _ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_paths(\n\u001b[1;32m    214\u001b[0m     \u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexperiment\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.json\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    215\u001b[0m )\n\u001b[1;32m    217\u001b[0m metadata[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mis_child\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_child\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/experiments/stagers/sce.py:390\u001b[0m, in \u001b[0;36mSingleCellExperimentStager._make_metadata\u001b[0;34m(self, update_info)\u001b[0m\n\u001b[1;32m    384\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m expt:\n\u001b[1;32m    385\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PYDSDBError(\n\u001b[1;32m    386\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExperiment: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_descriptor\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m metadata could not be \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    387\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparsed!\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    388\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mexpt\u001b[39;00m\n\u001b[0;32m--> 390\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mdict()\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/experiments/stagers/sce.py:324\u001b[0m, in \u001b[0;36mSingleCellExperimentStager._update_metadata\u001b[0;34m(self, metadata, update_info)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_update_metadata\u001b[39m(\n\u001b[1;32m    306\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    307\u001b[0m     metadata: SingleCellExperimentUp,\n\u001b[1;32m    308\u001b[0m     update_info: ExperimentUpdateInfo,\n\u001b[1;32m    309\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m SingleCellExperimentUp:\n\u001b[1;32m    310\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Update the contents of the metadata.\u001b[39;00m\n\u001b[1;32m    311\u001b[0m \n\u001b[1;32m    312\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    322\u001b[0m \u001b[38;5;124;03m        The updated metadata.\u001b[39;00m\n\u001b[1;32m    323\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 324\u001b[0m     metadata\u001b[38;5;241m.\u001b[39msummarized_experiment \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_stage_se_resources\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    325\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msummarized_experiment\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    326\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupdate_info\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43massay_nums\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    327\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    329\u001b[0m     metadata\u001b[38;5;241m.\u001b[39msingle_cell_experiment \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_stage_sce_resources(\n\u001b[1;32m    330\u001b[0m         metadata\u001b[38;5;241m.\u001b[39msingle_cell_experiment,\n\u001b[1;32m    331\u001b[0m         update_info\u001b[38;5;241m.\u001b[39malt_nums,\n\u001b[1;32m    332\u001b[0m     )\n\u001b[1;32m    334\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(metadata, SingleCellAccessibilityReferenceUp):\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/experiments/stagers/se.py:148\u001b[0m, in \u001b[0;36mSummarizedExperimentStagingMethods._stage_se_resources\u001b[0;34m(self, metadata, assay_nums)\u001b[0m\n\u001b[1;32m    143\u001b[0m metadata\u001b[38;5;241m.\u001b[39massays \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_combine_named_resources(\n\u001b[1;32m    144\u001b[0m     metadata\u001b[38;5;241m.\u001b[39massays, assays\n\u001b[1;32m    145\u001b[0m )\n\u001b[1;32m    147\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_artifact\u001b[38;5;241m.\u001b[39mcolData, DataFrame):\n\u001b[0;32m--> 148\u001b[0m     metadata\u001b[38;5;241m.\u001b[39mcolumn_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_child_resource_stager\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    149\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_data_frame_stager\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    150\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_artifact\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolData\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    151\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_descriptor\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m -> Column Data\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    152\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_staging_dir\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoinpath\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcoldata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    153\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstage\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    155\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PYDSDBError(\n\u001b[1;32m    156\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mStaging does not currently support \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mBiocFrame\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m in \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    157\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexperiment: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_descriptor\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m -> Column Data\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    158\u001b[0m     )\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/resources/generic.py:189\u001b[0m, in \u001b[0;36mResourceStager.stage\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Stage a resource.\u001b[39;00m\n\u001b[1;32m    174\u001b[0m \n\u001b[1;32m    175\u001b[0m \u001b[38;5;124;03mParameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    183\u001b[0m \u001b[38;5;124;03m    The `Resource` of the staged resource.\u001b[39;00m\n\u001b[1;32m    184\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    185\u001b[0m resource_path, relative_path, metadata_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_paths(\n\u001b[1;32m    186\u001b[0m     name, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_default_name, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_extension\n\u001b[1;32m    187\u001b[0m )\n\u001b[0;32m--> 189\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_write_resource\u001b[49m\u001b[43m(\u001b[49m\u001b[43mresource_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    191\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcheck_file_size(resource_path)\n\u001b[1;32m    193\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_write_metadata(\n\u001b[1;32m    194\u001b[0m     metadata_path,\n\u001b[1;32m    195\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_metadata(resource_path, relative_path),\n\u001b[1;32m    196\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_schema_dir\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_schema_name()\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    197\u001b[0m )\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/resources/stagers/csv_data_frame.py:79\u001b[0m, in \u001b[0;36mCSVDataFrameStager._write_resource\u001b[0;34m(self, resource_path)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_write_resource\u001b[39m(\u001b[38;5;28mself\u001b[39m, resource_path: Path) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m     72\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Write out a resource.\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \n\u001b[1;32m     74\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     77\u001b[0m \u001b[38;5;124;03m        The path to the resource file to write to.\u001b[39;00m\n\u001b[1;32m     78\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 79\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_column_resources \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_column_resources\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_artifact\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     81\u001b[0m     other_columns \u001b[38;5;241m=\u001b[39m {  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m     82\u001b[0m         column\u001b[38;5;241m.\u001b[39mname: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_artifact[column\u001b[38;5;241m.\u001b[39mname]\n\u001b[1;32m     83\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m column \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_column_resources\n\u001b[1;32m     84\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m column\u001b[38;5;241m.\u001b[39mresource \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m     85\u001b[0m     }\n\u001b[1;32m     86\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m other_columns\u001b[38;5;241m.\u001b[39mkeys():\n",
      "File \u001b[0;32m/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/pydsdb/upload/resources/stagers/data_frame_type_stager.py:96\u001b[0m, in \u001b[0;36mDataFrameTypeStager._make_column_resources\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m     91\u001b[0m value_types \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(\u001b[38;5;28mtype\u001b[39m(value) \u001b[38;5;28;01mfor\u001b[39;00m value \u001b[38;5;129;01min\u001b[39;00m column) \u001b[38;5;241m-\u001b[39m \u001b[38;5;28mset\u001b[39m(  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m     92\u001b[0m     [\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28;01mNone\u001b[39;00m)]\n\u001b[1;32m     93\u001b[0m )\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(value_types) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m---> 96\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m PYDSDBError(\n\u001b[1;32m     97\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFor column: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m with row \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mobject\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m dtype \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     98\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrow values have more than one type (excluding \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     99\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNone\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m): \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mvalue_types\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, this is not supported.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    100\u001b[0m     )\n\u001b[1;32m    102\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mbool\u001b[39m \u001b[38;5;129;01min\u001b[39;00m value_types \u001b[38;5;129;01mor\u001b[39;00m bool_ \u001b[38;5;129;01min\u001b[39;00m value_types:\n\u001b[1;32m    103\u001b[0m     column_resources\u001b[38;5;241m.\u001b[39mappend(\n\u001b[1;32m    104\u001b[0m         ColumnsField(name\u001b[38;5;241m=\u001b[39mname, \u001b[38;5;28mtype\u001b[39m\u001b[38;5;241m=\u001b[39mColumnTypes\u001b[38;5;241m.\u001b[39mBOOLEAN)\n\u001b[1;32m    105\u001b[0m     )\n",
      "\u001b[0;31mPYDSDBError\u001b[0m: For column: 'DemuxType_hashing' with row 'object' dtype row values have more than one type (excluding 'None'): '{<class 'float'>, <class 'str'>}', this is not supported."
     ]
    }
   ],
   "source": [
    "if DS_test==False:\n",
    "    dsid, version = DS.update_dataset(New_experiment, adata_updated, adatas_updated, test=False)\n",
    "else:    \n",
    "    dsid, version = DS.update_another_dataset(test_DatasetID, True, 1, New_experiment, adata_updated, adatas_updated, test=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('test-DS000004801', '2')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dsid, version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Retrieve back the dataset for checking purpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pydsdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "17:43:19 -> Collating dataset: 'test-DS000004801' metadata.\n",
      "\n",
      "17:43:23 -> Reading: 'test-DS000004801'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Column Data -> Data Frame: 100%|████████████████████████████████████████| 102/102 [00:00<00:00, 22.9kB/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "17:43:23 -> Reading: 'test-DS000004801 -> Experiment: main'.\n",
      "\n",
      "17:43:23 -> Reading: 'test-DS000004801 -> Experiment: main -> Alt. Exp.: crispr'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Alt. Exp.: crispr -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 243kB/s]\n",
      "Alt. Exp.: crispr -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 4.64M/4.64M [00:18<00:00, 248kB/s] \n",
      "Alt. Exp.: crispr -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 164k/164k [00:00<00:00, 186kB/s]  \n",
      "Alt. Exp.: crispr -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 51.8k/51.8k [00:00<00:00, 219kB/s]\n",
      "Experiment: main -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 200kB/s]\n",
      "Experiment: main -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 303M/303M [21:14<00:00, 238kB/s]    \n",
      "Experiment: main -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 743k/743k [00:02<00:00, 274kB/s]  \n",
      "Experiment: main -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 408k/408k [00:01<00:00, 205kB/s]  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "18:05:06 -> Reading: 'test-DS000004801 -> Experiment: raw_qc'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Experiment: raw_qc -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 379kB/s]\n",
      "Experiment: raw_qc -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 207M/207M [14:01<00:00, 246kB/s]    \n",
      "Experiment: raw_qc -> Assay: X -> Sparse Matrix: 100%|████████████████████████████████████████| 310M/310M [03:21<00:00, 1.54MB/s]   \n",
      "Experiment: raw_qc -> Column Data -> Data Frame -> Column: Sample -> Data Frame: 100%|████████████████████████████████████████| 57.0/57.0 [00:00<00:00, 19.1kB/s]\n",
      "Experiment: raw_qc -> Column Data -> Data Frame -> Column: Barcode -> Data Frame: 100%|████████████████████████████████████████| 105k/105k [00:00<00:00, 2.17MB/s]\n",
      "Experiment: raw_qc -> Column Data -> Data Frame -> Column: demux_type -> Data Frame: 100%|████████████████████████████████████████| 39.0/39.0 [00:00<00:00, 11.7kB/s]\n",
      "Experiment: raw_qc -> Column Data -> Data Frame -> Column: assignment -> Data Frame: 100%|████████████████████████████████████████| 58.0/58.0 [00:00<00:00, 17.2kB/s]\n",
      "Experiment: raw_qc -> Column Data -> Data Frame -> Column: phase -> Data Frame: 100%|████████████████████████████████████████| 40.0/40.0 [00:00<00:00, 11.8kB/s]\n",
      "Experiment: raw_qc -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 1.67M/1.67M [00:00<00:00, 15.8MB/s]\n",
      "Experiment: raw_qc -> Row Data -> Data Frame -> Column: Symbol -> Data Frame: 100%|████████████████████████████████████████| 76.1k/76.1k [00:00<00:00, 24.2MB/s]\n",
      "Experiment: raw_qc -> Row Data -> Data Frame -> Column: Type -> Data Frame: 100%|████████████████████████████████████████| 47.0/47.0 [00:00<00:00, 20.0kB/s]\n",
      "Experiment: raw_qc -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 886k/886k [00:00<00:00, 68.7MB/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "18:22:35 -> Reading: 'test-DS000004801 -> Experiment: crispr_hashing_harmonized'.\n",
      "\n",
      "18:22:35 -> Reading: 'test-DS000004801 -> Experiment: crispr_hashing_harmonized -> Alt. Exp.: crispr'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Alt. Exp.: crispr -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 173kB/s]\n",
      "Alt. Exp.: crispr -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 4.64M/4.64M [00:01<00:00, 3.93MB/s]\n",
      "Alt. Exp.: crispr -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 164k/164k [00:00<00:00, 32.3MB/s]\n",
      "Alt. Exp.: crispr -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 51.8k/51.8k [00:00<00:00, 1.33MB/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "18:22:37 -> Reading: 'test-DS000004801 -> Experiment: crispr_hashing_harmonized -> Alt. Exp.: hashing'.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Alt. Exp.: hashing -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 232kB/s]\n",
      "Alt. Exp.: hashing -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 595k/595k [00:00<00:00, 1.46MB/s]\n",
      "Alt. Exp.: hashing -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 164k/164k [00:00<00:00, 3.38MB/s]\n",
      "Alt. Exp.: hashing -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 59.0/59.0 [00:00<00:00, 17.2kB/s]\n",
      "Experiment: crispr_hashing_harmonized -> Reduced Dimensions -> NDArrays: 100%|████████████████████████████████████████| 800/800 [00:00<00:00, 232kB/s]\n",
      "Experiment: crispr_hashing_harmonized -> Assay: counts -> Sparse Matrix: 100%|████████████████████████████████████████| 303M/303M [00:37<00:00, 8.18MB/s] \n",
      "Experiment: crispr_hashing_harmonized -> Column Data -> Data Frame: 100%|████████████████████████████████████████| 644k/644k [00:00<00:00, 23.5MB/s]\n",
      "Experiment: crispr_hashing_harmonized -> Row Data -> Data Frame: 100%|████████████████████████████████████████| 408k/408k [00:00<00:00, 97.2MB/s]\n",
      "Sample Mapping: 100%|████████████████████████████████████████| 725k/725k [00:00<00:00, 24.1MB/s]\n",
      "/apps/user/gpy/envs/dev/GPy39/lib/python3.9/site-packages/multiassayexperiment/MultiAssayExperiment.py:112: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.\n",
      "  for group, rows in agroups:\n"
     ]
    }
   ],
   "source": [
    "dss = pydsdb.get_dataset(dsid, dev=DEV, version=version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'.internal': {'metadata': '{\"dataset\": {\"authors\": [\"SG\"], \"column_data\": {\"resource\": {\"path\": \"sample_data/simple.csv.gz\", \"type\": \"local\"}}, \"creation_date\": \"2023-04-05T19:11:25+00:00\", \"description\": \"Siavash test upload; this should be gone\", \"experiments\": [{\"name\": \"main\", \"resource\": {\"path\": \"experiment-1/experiment.json\", \"type\": \"local\"}}, {\"name\": \"raw_qc\", \"resource\": {\"path\": \"experiment-2/experiment.json\", \"type\": \"local\"}}, {\"name\": \"crispr_hashing_harmonized\", \"resource\": {\"path\": \"experiment-3/experiment.json\", \"type\": \"local\"}}], \"sample_mapping\": {\"resource\": {\"path\": \"sample_mapping/simple.csv.gz\", \"type\": \"local\"}}, \"title\": \"Siavash test upload\"}, \"is_child\": false, \"path\": \"dataset.json\", \"$schema\": \"dataset/v2.json\"}',\n",
       "  'paths': ['dataset.json',\n",
       "   'sample_data/simple.csv.gz',\n",
       "   'sample_mapping/simple.csv.gz'],\n",
       "  'column_data':                                                                            samples\n",
       "  unknown_sample_crispr_hashing_harmonized  unknown_sample_crispr_hashing_harmonized\n",
       "  unknown_sample_raw_qc                                        unknown_sample_raw_qc\n",
       "  unknown_sample_main                                            unknown_sample_main,\n",
       "  'sample_mapping':                            assay                                   primary  \\\n",
       "  0      crispr_hashing_harmonized  unknown_sample_crispr_hashing_harmonized   \n",
       "  1      crispr_hashing_harmonized  unknown_sample_crispr_hashing_harmonized   \n",
       "  2      crispr_hashing_harmonized  unknown_sample_crispr_hashing_harmonized   \n",
       "  3      crispr_hashing_harmonized  unknown_sample_crispr_hashing_harmonized   \n",
       "  4      crispr_hashing_harmonized  unknown_sample_crispr_hashing_harmonized   \n",
       "  ...                          ...                                       ...   \n",
       "  37066                       main                       unknown_sample_main   \n",
       "  37067                       main                       unknown_sample_main   \n",
       "  37068                       main                       unknown_sample_main   \n",
       "  37069                       main                       unknown_sample_main   \n",
       "  37070                       main                       unknown_sample_main   \n",
       "  \n",
       "                                           colname  \n",
       "  0      SAM24424070_Replicate1-AAACCCAAGACGTCGA-1  \n",
       "  1      SAM24424070_Replicate1-AAACCCAAGAGATTCA-1  \n",
       "  2      SAM24424070_Replicate1-AAACCCAAGCCAGACA-1  \n",
       "  3      SAM24424070_Replicate1-AAACCCAAGCGATGCA-1  \n",
       "  4      SAM24424070_Replicate1-AAACCCAAGGATATGT-1  \n",
       "  ...                                          ...  \n",
       "  37066  SAM24424071_Replicate1-TTTGTTGTCGCCGAAC-1  \n",
       "  37067  SAM24424071_Replicate1-TTTGTTGTCGCGAAGA-1  \n",
       "  37068  SAM24424071_Replicate1-TTTGTTGTCGCTGTCT-1  \n",
       "  37069  SAM24424071_Replicate1-TTTGTTGTCGTCACCT-1  \n",
       "  37070  SAM24424071_Replicate1-TTTGTTGTCTGGGTCG-1  \n",
       "  \n",
       "  [100044 rows x 3 columns],\n",
       "  'exp_nums': {'main': 1, 'raw_qc': 2, 'crispr_hashing_harmonized': 3},\n",
       "  'permissions_info': {'owners': ['ghaffars'],\n",
       "   'read_access': 'public',\n",
       "   'scope': 'project',\n",
       "   'write_access': 'owners'}}}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        Class MultiAssayExperiment with 3 experiments and 3 samples\n",
      "            experiments: \n",
      "                [\"main: \\n        Class SingleCellExperiment with 36603 features and 37071 cells\\n          mainExperimentName: None\\n          assays: ['counts']\\n          features: Index(['ID', 'Symbol', 'Type'], dtype='object')\\n          cell metadata: Index(['Sample', 'Barcode', 'demux_type', 'assignment', 'pBarcode_assignment'], dtype='object')\\n          reduced dimensions: []\\n          alternative experiments: ['crispr']\\n        \", \"raw_qc: \\n        Class SingleCellExperiment with 26987 features and 25902 cells\\n          mainExperimentName: None\\n          assays: ['counts', 'X']\\n          features: Index(['ID', 'Symbol', 'Type', 'mt', 'ribo', 'n_cells_by_counts',\\n       'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells'],\\n      dtype='object')\\n          cell metadata: Index(['Sample', 'Barcode', 'demux_type', 'assignment', 'n_genes_by_counts',\\n       'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo',\\n       'pct_counts_ribo', 'n_genes', 'phase', 'G2M_score', 'S_score'],\\n      dtype='object')\\n          reduced dimensions: []\\n          alternative experiments: None\\n        \", \"crispr_hashing_harmonized: \\n        Class SingleCellExperiment with 36603 features and 37071 cells\\n          mainExperimentName: None\\n          assays: ['counts']\\n          features: Index(['ID', 'Symbol', 'Type'], dtype='object')\\n          cell metadata: Index(['Sample', 'Barcode', 'DemuxType_crispr', 'DemuxAssignment_crispr',\\n       'DemuxType_hashing', 'DemuxAssignment_hashing', 'Protocol'],\\n      dtype='object')\\n          reduced dimensions: []\\n          alternative experiments: ['crispr', 'hashing']\\n        \"]\n",
      "        \n"
     ]
    }
   ],
   "source": [
    "print(dss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_2, adatas_2 = dss.experiments[\"main\"].toAnnData(alts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37071 × 36603\n",
       "    obs: 'Sample', 'Barcode', 'demux_type', 'assignment', 'pBarcode_assignment'\n",
       "    var: 'ID', 'Symbol', 'Type'\n",
       "    uns: '.internal'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'crispr': AnnData object with n_obs × n_vars = 37071 × 21091\n",
       "     uns: '.internal'\n",
       "     layers: 'counts'}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_3, adatas_3 = dss.experiments[\"crispr_hashing_harmonized\"].toAnnData(alts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 37071 × 36603\n",
       "    obs: 'Sample', 'Barcode', 'DemuxType_crispr', 'DemuxAssignment_crispr', 'DemuxType_hashing', 'DemuxAssignment_hashing', 'Protocol'\n",
       "    var: 'ID', 'Symbol', 'Type'\n",
       "    uns: '.internal'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'crispr': AnnData object with n_obs × n_vars = 37071 × 21091\n",
       "     uns: '.internal'\n",
       "     layers: 'counts',\n",
       " 'hashing': AnnData object with n_obs × n_vars = 37071 × 8\n",
       "     uns: '.internal'\n",
       "     layers: 'counts'}"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas_3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gpydev39",
   "language": "python",
   "name": "gpydev39"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
